---
title: "mutscan"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{mutscan}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(SummarizedExperiment)
library(mutscan)
```

## Read example data

```{r readTrans}
datadir <- system.file("extdata", package = "mutscan")
dir(datadir)
transInput <- digestFastqs(experimentType = "trans", fastqForward = file.path(datadir, "transInput_1.fastq.gz"),
                           fastqReverse = file.path(datadir, "transInput_2.fastq.gz"),
                           skipForward = 1, skipReverse = 1, umiLengthForward = 10, 
                           umiLengthReverse = 8, constantLengthForward = 18,
                           constantLengthReverse = 20, variableLengthForward = 96,
                           variableLengthReverse = 96, constantForward = "AACCGGAGGAGGGAGCTG",
                           constantReverse = "GAAAAAGGAAGCTGGAGAGA", adapterForward = "GGAAGAGCACACGTC", 
                           adapterReverse = "GGAAGAGCGTCGTGT", 
                           avePhredMin = 20, variableNMax = 0, umiNMax = 0,
                           wildTypeForward = "ACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA", 
                           wildTypeReverse = "ATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTT", 
                           nbrMutatedCodonsMax = 1, forbiddenMutatedCodon = "NNW",verbose = FALSE)
transInput$parameters
transInput$filterSummary
head(transInput$summaryTable)
transInput$errorStatistics[rowSums(transInput$errorStatistics[, -1]) != 0, ]
transInput$experimentType
```

## Estimate sequencing error rate from constant sequences

```{r errStats}
(propErrorsConstantF <- sum(transInput$errorStatistics$nbrMismatchForward)/
   (transInput$parameters$constantLengthForward * transInput$filterSummary$nbrRetained))
(propErrorsConstantR <- sum(transInput$errorStatistics$nbrMismatchReverse)/
   (transInput$parameters$constantLengthReverse * transInput$filterSummary$nbrRetained))
```

## Combine samples (will count occurrences of each combination of variants) 

```{r combine}
# se <- summarizeExperiment(x = list(sample1 = transInput,
#                                    sample2 = transInput),
#                           coldata = data.frame(Name = c("sample1", "sample2"),
#                                                Condition = c("input", "enriched")))
# head(assays(se)$counts)
```

