---
title: "mutscan"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{mutscan}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(mutscan)
```

## Read example data

```{r readTrans}
datadir <- system.file("extdata", package = "mutscan")
dir(datadir)
transInput <- readFastqs(experimentType = "trans", fastqForward = file.path(datadir, "transInput_1.fastq.gz"),
                         fastqReverse = file.path(datadir, "transInput_2.fastq.gz"),
                         skipForward = 1, skipReverse = 1, umiLengthForward = 10, 
                         umiLengthReverse = 8, constantLengthForward = 18,
                         constantLengthReverse = 20, adapterForward = "GGAAGAGCACACGTC", 
                         adapterReverse = "GGAAGAGCGTCGTGT", verbose = TRUE)
colData(transInput)
```

## Filter reads

```{r filterReads}
transInputFiltered <- filterReads(transInput, avePhredMin = 20, 
                                  variableNMax = 0, umiNMax = 0, 
                                  wildTypeForward = "ACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA",
                                  wildTypeReverse = "ATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTT", 
                                  nbrMutatedCodonsMax = 1, forbiddenMutatedCodon = "NNW")
colData(transInputFiltered)
```

## Estimate sequencing error rate from constant sequences

```{r errStats}
# user-provided expected sequences
# errStats <- calcErrorStats(transInputFiltered, expectedForward = "AACCGGAGGAGGGAGCTG", 
#                            expectedReverse = "GAAAAAGGAAGCTGGAGAGA", verbose = TRUE)

# auto-detect expected sequences
errStats <- calcErrorStats(transInputFiltered, verbose = TRUE)
errStats$propErrorsConstantF
errStats$propErrorsConstantR
```

## Combine samples (will count occurrences of each combination of variants) 

```{r combine}
se <- summarizeExperiment(x = list(sample1 = transInputFiltered,
                                   sample2 = transInputFiltered),
                          meta = data.frame(Name = c("sample1", "sample2"),
                                            Condition = c("input", "enriched")))
head(assays(se)$counts)
```

