<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Collapse mutants by similarity — collapseMutantsBySimilarity • mutscan</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Collapse mutants by similarity — collapseMutantsBySimilarity"><meta name="description" content='These functions can be used to collapse variants, either by similarity or
according to a pre-defined grouping. The functions collapseMutants and
collapseMutantsByAA assume that a grouping variable is available as a
column in rowData(se) (collapseMutantsByAA is a convenience
function for the case when this column is "mutantNameAA", and is provided
for backwards compatibility). The collapseMutantsBySimilarity will
generate the grouping variable based on user-provided thresholds on the
sequence similarity (defined by the Hamming distance), and subsequently
collapse based on the derived grouping.'><meta property="og:description" content='These functions can be used to collapse variants, either by similarity or
according to a pre-defined grouping. The functions collapseMutants and
collapseMutantsByAA assume that a grouping variable is available as a
column in rowData(se) (collapseMutantsByAA is a convenience
function for the case when this column is "mutantNameAA", and is provided
for backwards compatibility). The collapseMutantsBySimilarity will
generate the grouping variable based on user-provided thresholds on the
sequence similarity (defined by the Hamming distance), and subsequently
collapse based on the derived grouping.'></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">mutscan</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/mutscan.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Collapse mutants by similarity</h1>

      <div class="d-none name"><code>collapseMutantsBySimilarity.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>These functions can be used to collapse variants, either by similarity or
according to a pre-defined grouping. The functions <code>collapseMutants</code> and
<code>collapseMutantsByAA</code> assume that a grouping variable is available as a
column in <code>rowData(se)</code> (<code>collapseMutantsByAA</code> is a convenience
function for the case when this column is "mutantNameAA", and is provided
for backwards compatibility). The <code>collapseMutantsBySimilarity</code> will
generate the grouping variable based on user-provided thresholds on the
sequence similarity (defined by the Hamming distance), and subsequently
collapse based on the derived grouping.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">collapseMutantsBySimilarity</span><span class="op">(</span></span>
<span>  <span class="va">se</span>,</span>
<span>  <span class="va">assayName</span>,</span>
<span>  scoreMethod <span class="op">=</span> <span class="st">"rowSum"</span>,</span>
<span>  sequenceCol <span class="op">=</span> <span class="st">"sequence"</span>,</span>
<span>  collapseMaxDist <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  collapseMinScore <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  collapseMinRatio <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">collapseMutantsByAA</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">collapseMutants</span><span class="op">(</span><span class="va">se</span>, <span class="va">nameCol</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-se">se<a class="anchor" aria-label="anchor" href="#arg-se"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code>
generated by <code><a href="summarizeExperiment.html">summarizeExperiment</a></code></p></dd>


<dt id="arg-assayname">assayName<a class="anchor" aria-label="anchor" href="#arg-assayname"></a></dt>
<dd><p>The name of the assay that will be used to calculate a
"score" (typically derived from the read counts) for each variant.</p></dd>


<dt id="arg-scoremethod">scoreMethod<a class="anchor" aria-label="anchor" href="#arg-scoremethod"></a></dt>
<dd><p>Character scalar giving the approach used to calculate
ranking scores from the assay defined by <code>assayName</code>. Currently,
this can be one of <code>"rowSum"</code> or <code>"rowMean"</code>. All filtering
criteria will be applied to these scores.</p></dd>


<dt id="arg-sequencecol">sequenceCol<a class="anchor" aria-label="anchor" href="#arg-sequencecol"></a></dt>
<dd><p>Character scalar giving the name of the column in
<code>rowData(se)</code> that contains the nucleotide sequence of the
variants.</p></dd>


<dt id="arg-collapsemaxdist">collapseMaxDist<a class="anchor" aria-label="anchor" href="#arg-collapsemaxdist"></a></dt>
<dd><p>Numeric scalar defining the tolerance for collapsing
similar sequences. If the value is in [0, 1), it defines the maximal
Hamming distance in terms of a fraction of sequence length:
(<code>round(collapseMaxDist * nchar(sequence))</code>).
A value greater or equal to 1 is rounded and directly used as the maximum
allowed Hamming distance. Note that sequences can only be
collapsed if they are all of the same length.</p></dd>


<dt id="arg-collapseminscore">collapseMinScore<a class="anchor" aria-label="anchor" href="#arg-collapseminscore"></a></dt>
<dd><p>Numeric scalar, indicating the minimum score for the
sequence to be considered for collapsing with similar sequences.</p></dd>


<dt id="arg-collapseminratio">collapseMinRatio<a class="anchor" aria-label="anchor" href="#arg-collapseminratio"></a></dt>
<dd><p>Numeric scalar. During collapsing of
similar sequences, a low-frequency sequence will be collapsed
with a higher-frequency sequence only if the ratio between the
high-frequency and the low-frequency scores is at least this
high. The default value of 0 indicates that no such check is performed.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical, whether to print progress messages.</p></dd>


<dt id="arg-namecol">nameCol<a class="anchor" aria-label="anchor" href="#arg-namecol"></a></dt>
<dd><p>A character scalar providing the column of
<code>rowData(se)</code> that contains the amino acid mutant names (that will
be the new row names).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code> where
    counts have been aggregated by the mutated amino acid(s).</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Charlotte Soneson, Michael Stadler</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"GSE102901_cis_se.rds"</span>,</span></span>
<span class="r-in"><span>                          package <span class="op">=</span> <span class="st">"mutscan"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">200</span>, <span class="op">]</span></span></span>
<span class="r-in"><span><span class="co">## The rows of this object correspond to individual codon variants</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 200   6</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "f.0.WT"  "f.1.AAC" "f.1.AAG" "f.1.ACC" "f.1.ACG" "f.1.AGC"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Collapse by amino acid</span></span></span>
<span class="r-in"><span><span class="va">sec</span> <span class="op">&lt;-</span> <span class="fu">collapseMutantsByAA</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">## The rows of the collapsed object correspond to amino acid variants</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sec</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 128   6</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sec</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "f.0.WT" "f.1.*"  "f.1.A"  "f.1.C"  "f.1.D"  "f.1.E" </span>
<span class="r-in"><span><span class="co">## The mutantName column contains the individual codon variants that were </span></span></span>
<span class="r-in"><span><span class="co">## collapsed</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sec</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 6 rows and 19 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        mutantNameAA               sequence             mutantName</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         &lt;character&gt;            &lt;character&gt;            &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.0.WT       f.0.WT ACCGATACACTCCAAGCGGA.. f.0.WT,f.1.ACC,f.1.A..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.*         f.1.* TAGGATACACTCCAAGCGGA..                f.1.TAG</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.A         f.1.A GCCGATACACTCCAAGCGGA..        f.1.GCC,f.1.GCG</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.C         f.1.C TGCGATACACTCCAAGCGGA..                f.1.TGC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.D         f.1.D GACGATACACTCCAAGCGGA..                f.1.GAC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.E         f.1.E GAGGATACACTCCAAGCGGA..                f.1.GAG</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                mutantNameBase     mutantNameBaseHGVS        mutantNameCodon</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   &lt;character&gt;            &lt;character&gt;            &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.0.WT f.0.WT,f.3.C,f.3.G,f.. f:c,f:c.30A&gt;G,f:c.31.. f.0.WT,f.1.ACC,f.1.A..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.*       f.1.T_f.2.A_f.3.G       f:c.1_3delinsTAG                f.1.TAG</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.A  f.1.G_f.3.C,f.1.G_f... f:c.1_3delinsGCC,f:c..        f.1.GCC,f.1.GCG</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.C       f.1.T_f.2.G_f.3.C       f:c.1_3delinsTGC                f.1.TGC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.D       f.1.G_f.2.A_f.3.C       f:c.1_3delinsGAC                f.1.GAC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.E       f.1.G_f.2.A_f.3.G       f:c.1_3delinsGAG                f.1.GAG</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        mutantNameAAHGVS             sequenceAA mutationTypes nbrMutBases</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             &lt;character&gt;            &lt;character&gt;   &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.0.WT              f:p TDTLQAETDQLEDEKSALQT..        silent       0,1,2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.*       f:p.(Thr1*) *DTLQAETDQLEDEKSALQT..          stop           3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.A     f:p.(Thr1Ala) ADTLQAETDQLEDEKSALQT.. nonsynonymous           2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.C     f:p.(Thr1Cys) CDTLQAETDQLEDEKSALQT.. nonsynonymous           3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.D     f:p.(Thr1Asp) DDTLQAETDQLEDEKSALQT.. nonsynonymous           3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.E     f:p.(Thr1Glu) EDTLQAETDQLEDEKSALQT.. nonsynonymous           3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        nbrMutCodons   nbrMutAAs  varLengths minNbrMutBases minNbrMutCodons</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         &lt;character&gt; &lt;character&gt; &lt;character&gt;      &lt;integer&gt;       &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.0.WT          0,1           0          96              0               0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.*             1           1          96              3               1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.A             1           1          96              2               1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.C             1           1          96              3               1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.D             1           1          96              3               1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.E             1           1          96              3               1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        minNbrMutAAs maxNbrMutBases maxNbrMutCodons maxNbrMutAAs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           &lt;integer&gt;      &lt;integer&gt;       &lt;integer&gt;    &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.0.WT            0              2               1            0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.*             1              3               1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.A             1              2               1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.C             1              3               1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.D             1              3               1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.E             1              3               1            1</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Collapse similar sequences</span></span></span>
<span class="r-in"><span><span class="va">sec2</span> <span class="op">&lt;-</span> <span class="fu">collapseMutantsBySimilarity</span><span class="op">(</span></span></span>
<span class="r-in"><span>    se <span class="op">=</span> <span class="va">se</span>, assayName <span class="op">=</span> <span class="st">"counts"</span>, scoreMethod <span class="op">=</span> <span class="st">"rowSum"</span>,</span></span>
<span class="r-in"><span>    sequenceCol <span class="op">=</span> <span class="st">"sequence"</span>, collapseMaxDist <span class="op">=</span> <span class="fl">2</span>,</span></span>
<span class="r-in"><span>    collapseMinScore <span class="op">=</span> <span class="fl">0</span>, collapseMinRatio <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> start collapsing sequences (tolerance: 2)...done (reduced from 200 to 12)</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sec2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 12  6</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sec2</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "f.0.WT"   "f.1.GGG"  "f.1.TTC"  "f.10.GGG" "f.10.TCC" "f.11.GCC"</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sec2</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 6 rows and 20 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          collapseCol               sequence             mutantName</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;character&gt;            &lt;character&gt;            &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.0.WT        f.0.WT AACGATACACTCCAAGCGGA.. f.0.WT,f.1.AAC,f.1.A..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.GGG      f.1.GGG CAGGATACACTCCAAGCGGA.. f.1.CAG,f.1.CGC,f.1...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.TTC      f.1.TTC CACGATACACTCCAAGCGGA.. f.1.CAC,f.1.CTC,f.1...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.10.GGG    f.10.GGG ACTGATACACTCCAAGCGGA.. f.10.ACG,f.10.AGC,f...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.10.TCC    f.10.TCC ACTGATACACTCCAAGCGGA.. f.10.ACC,f.10.ATC,f...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.11.GCC    f.11.GCC ACTGATACACTCCAAGCGGA.. f.11.ACC,f.11.AGC,f...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  mutantNameBase     mutantNameBaseHGVS        mutantNameCodon</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     &lt;character&gt;            &lt;character&gt;            &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.0.WT   f.0.WT,f.1.C_f.3.C,f.. f:c,f:c.1_3delinsCCC.. f.0.WT,f.1.AAC,f.1.A..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.GGG  f.1.C_f.2.A_f.3.G,f... f:c.1_3delinsCAG,f:c.. f.1.CAG,f.1.CGC,f.1...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.TTC  f.1.C_f.2.A_f.3.C,f... f:c.1_3delinsCAC,f:c.. f.1.CAC,f.1.CTC,f.1...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.10.GGG f.28.A_f.29.C_f.30.G.. f:c.28_30delinsACG,f.. f.10.ACG,f.10.AGC,f...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.10.TCC f.28.A_f.29.C_f.30.C.. f:c.28_30delinsACC,f.. f.10.ACC,f.10.ATC,f...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.11.GCC f.31.A_f.32.C_f.33.C.. f:c.31_33delinsACC,f.. f.11.ACC,f.11.AGC,f...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    mutantNameAA       mutantNameAAHGVS             sequenceAA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     &lt;character&gt;            &lt;character&gt;            &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.0.WT   f.0.WT,f.1.A,f.1.I,f.. f:p,f:p.(Asp13*),f:p.. ADTLQAETDQLEDEKSALQT..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.GGG  f.1.*,f.1.C,f.1.D,f... f:p.(Thr1*),f:p.(Thr.. *DTLQAETDQLEDEKSALQT..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.TTC  f.1.F,f.1.H,f.1.L,f... f:p.(Thr1His),f:p.(T.. FDTLQAETDQLEDEKSALQT..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.10.GGG f.10.A,f.10.C,f.10.G.. f:p.(Gln10Ala),f:p.(.. TDTLQAETDALEDEKSALQT..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.10.TCC f.10.F,f.10.I,f.10.S.. f:p.(Gln10Ile),f:p.(.. TDTLQAETDFLEDEKSALQT..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.11.GCC f.11.A,f.11.G,f.11.S.. f:p.(Leu11Ala),f:p.(.. TDTLQAETDQAEDEKSALQT..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   mutationTypes nbrMutBases nbrMutCodons   nbrMutAAs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     &lt;character&gt; &lt;character&gt;  &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.0.WT   nonsynonymous,silent..       0,1,2          0,1         0,1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.GGG      nonsynonymous,stop           3            1           1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.TTC           nonsynonymous           3            1           1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.10.GGG          nonsynonymous           3            1           1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.10.TCC          nonsynonymous           3            1           1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.11.GCC          nonsynonymous           3            1           1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           varLengths minNbrMutBases minNbrMutCodons minNbrMutAAs maxNbrMutBases</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;character&gt;      &lt;integer&gt;       &lt;integer&gt;    &lt;integer&gt;      &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.0.WT            96              0               0            0              2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.GGG           96              3               1            1              3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.TTC           96              3               1            1              3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.10.GGG          96              3               1            1              3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.10.TCC          96              3               1            1              3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.11.GCC          96              3               1            1              3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          maxNbrMutCodons maxNbrMutAAs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                &lt;integer&gt;    &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.0.WT                 1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.GGG                1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.TTC                1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.10.GGG               1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.10.TCC               1            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.11.GCC               1            1</span>
<span class="r-in"><span><span class="co">## collapsed count matrix</span></span></span>
<span class="r-in"><span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sec2</span>, <span class="st">"counts"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          SRR5952435 SRR5952436 SRR5952437 SRR5952438 SRR5952439 SRR5952440</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.0.WT      3068572    3757468    3572269    6831160    6678754    6624525</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.GGG         621        726        503        190         98        208</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.1.TTC          97        115        109         27         18         23</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.10.GGG        470        511        371        173         84        181</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.10.TCC         36         82         32         11          7         20</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.11.GCC         61         74         42         13         10         14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.11.TAG        259        294        196         55         41         64</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.12.AGG         81         81         51         19         15         46</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.12.CCC        286        294        167         76         45         89</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.13.CGC        471        581        447        156         87        195</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.13.TTG        114        146         86         37         12         42</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> f.14.CGC        122        137        123         36         26         42</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Charlotte Soneson, Michael Stadler.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

