<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Process an experiment with multiple variable sequences — linkMultipleVariants • mutscan</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Process an experiment with multiple variable sequences — linkMultipleVariants"><meta name="description" content="This function enables the processing of data sets with multiple
variable sequences, which should potentially be handled in different
ways. For example, a barcode association experiment
with two variable sequences (the barcode and the biological variant)
that need to be processed differently, e.g. in terms of matching to
wildtype sequences or collapsing of similar sequences.
In contrast, while digestFastqs allow the specification
of multiple variable sequences (within each of the forward and reverse
reads), they will be concatenated and processed as a single unit."><meta property="og:description" content="This function enables the processing of data sets with multiple
variable sequences, which should potentially be handled in different
ways. For example, a barcode association experiment
with two variable sequences (the barcode and the biological variant)
that need to be processed differently, e.g. in terms of matching to
wildtype sequences or collapsing of similar sequences.
In contrast, while digestFastqs allow the specification
of multiple variable sequences (within each of the forward and reverse
reads), they will be concatenated and processed as a single unit."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">mutscan</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/mutscan.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Process an experiment with multiple variable sequences</h1>

      <div class="d-none name"><code>linkMultipleVariants.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function enables the processing of data sets with multiple
variable sequences, which should potentially be handled in different
ways. For example, a barcode association experiment
with two variable sequences (the barcode and the biological variant)
that need to be processed differently, e.g. in terms of matching to
wildtype sequences or collapsing of similar sequences.
In contrast, while <code>digestFastqs</code> allow the specification
of multiple variable sequences (within each of the forward and reverse
reads), they will be concatenated and processed as a single unit.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">linkMultipleVariants</span><span class="op">(</span>combinedDigestParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-combineddigestparams">combinedDigestParams<a class="anchor" aria-label="anchor" href="#arg-combineddigestparams"></a></dt>
<dd><p>A named list of arguments to
<code>digestFastqs</code> for the combined ("naive") run.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments providing arguments to <code>digestFastqs</code>
for the separate runs (processing each variable sequence in turn).
Each argument must be a named list of arguments to <code>digestFastqs</code>.
In addition, arguments <code>collapseMaxDist</code>, <code>collapseMinScore</code>
and <code>collapseMinRatio</code> can be specified, and will be passed on
to <code>collapseMutantsBySimilarity</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list with the following elements:</p><ul><li><p>countAggregated - a <code>tibble</code> with columns corresponding to
    each of the variable sequences, and a column with the total observed
    read count for the combination.</p></li>
<li><p>convSeparate - a list of conversion tables from the respective
    separate runs.</p></li>
<li><p>outCombined - the <code>digestFastqs</code> output for the combined run.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>linkMultipleVariants will process the input in the following way:</p><ul><li><p>First, run <code>digestFastqs</code> with the parameters provided
    in <code>combinedDigestParams</code>. Typically, this will be a
    "naive" counting run, where the frequencies of all observed
    variants are tabulated. The variable sequences
    within the forward and reverse reads, respectively, will be
    processed as a single sequence.</p></li>
<li><p>Next, run <code>digestFastqs</code> with each of the additional
    parameter sets provided (<code>...</code>). Each of these should
    correspond to a single variable sequence from the combined
    run (i.e., if there are two Vs in the element specifications
    in the combined run, there should be two additional
    parameter sets provided, each corresponding to the
    processing of one variable sequence part). It is assumed
    that the order of the additional arguments correspond to the
    order of the variable sequences in the combined run, in such a way
    that if the variable sequences extracted in each of the separate
    runs are concatenated in the order that the parameter sets are
    provided to <code>linkMultipleVariants</code>, they will form the variable
    sequence extracted in the combined run.</p></li>
<li><p>The result of each of the separate runs is a 'conversion table',
    containing the final set of identified sequence variants as well
    as all individual sequences corresponding to each of them. This
    is then combined with the count table from the combined, "naive"
    run in order to create an aggregated count table. More precisely,
    each sequence in the combined run is split into the constituent
    variable sequences, and
    each variable sequence is then matched to the output from the right
    separate run, from which the final feature ID (mutant name, or
    collapsed sequence) will be extracted and used to replace the original
    sequence in the combined count table. Once all the matches are done,
    rows with NAs (where no match could be found in the separate run)
    are removed and the counts are aggregated across all identical
    combinations of variable sequences.</p></li>
</ul><p>In order to define the <code>elementsForward</code> and <code>elementsReverse</code>
arguments for the separate runs, a strategy that often works is to simply
copy the arguments from the combined run, and successively replace all
but one of the 'V's by 'S'. This will effectively process one variable
sequence at the time, while keeping all other elements of the reads
consistent (since this can affect e.g. filtering criteria). Note that
to process individual variable sequences in the reverse read, you also
need to swap the 'forward' and 'reverse' specifications (since
<code>digestFastqs</code> requires a forward read).</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Charlotte Soneson, Michael Stadler</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Charlotte Soneson, Michael Stadler.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

