<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Deep mutational scanning analysis with `mutscan` • mutscan</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Deep mutational scanning analysis with `mutscan`">
<meta property="og:description" content="mutscan">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mutscan</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.26</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/mutscan.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="mutscan_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Deep mutational scanning analysis with <code>mutscan</code>
</h1>
                        <h4 data-toc-skip class="author">Charlotte Soneson and Michael Stadler</h4>
            
            <h4 data-toc-skip class="date">21 January 2022</h4>
      
      
      <div class="hidden name"><code>mutscan.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">mutscan</span><span class="op">)</span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>mutscan</code> provides functionality for processing and visualizing Deep Mutational Scanning and other similar types of data, starting from single-end or paired-end FASTQ files. A broad range of library designs can be processed, encompassing both TRANS experiments (considering pairs of proteins) and CIS experiments (studying a single protein).</p>
<p>The figure below provides a high-level overview of the <code>mutscan</code> functionality, which will be described in more detail in the following sections.</p>
<div class="figure" style="text-align: center">
<img src="mutscan-overview.001.png" alt="Overview of the `mutscan` functionality. The `digestFastqs()` function processes each sequencing library separately, and generates an output object containing, among other things, the count table and a summary of the filtering steps. The `summarizeExperiment()` function takes one or more of these objects and combines them into a `SummarizedExperiment` object, that can then be used for downstream analysis such as plotting and statistical testing." width="100%"><p class="caption">
Overview of the <code>mutscan</code> functionality. The <code><a href="../reference/digestFastqs.html">digestFastqs()</a></code> function processes each sequencing library separately, and generates an output object containing, among other things, the count table and a summary of the filtering steps. The <code><a href="../reference/summarizeExperiment.html">summarizeExperiment()</a></code> function takes one or more of these objects and combines them into a <code>SummarizedExperiment</code> object, that can then be used for downstream analysis such as plotting and statistical testing.
</p>
</div>
</div>
<div class="section level2">
<h2 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h2>
<p>The <code>mutscan</code> package contains several small example FASTQ files representing different types of experiments:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">datadir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"mutscan"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="va">datadir</span><span class="op">)</span>
<span class="co">#&gt;  [1] "cisInput_1.fastq.gz"    "cisInput_2.fastq.gz"    "cisOutput_1.fastq.gz"  </span>
<span class="co">#&gt;  [4] "cisOutput_2.fastq.gz"   "GSE102901_cis_se.rds"   "leujunt0_1.fastq.gz"   </span>
<span class="co">#&gt;  [7] "leujunt0_2.fastq.gz"    "transInput_1.fastq.gz"  "transInput_2.fastq.gz" </span>
<span class="co">#&gt; [10] "transOutput_1.fastq.gz" "transOutput_2.fastq.gz"</span></code></pre></div>
<ul>
<li>
<code>transInput_{1,2}.fastq.gz</code>, <code>transOutput_{1,2}.fastq.gz</code> - data from TRANS experiment in <span class="citation">[@Diss2018]</span>. The forward and reverse reads correspond to the mutated <em>FOS</em> and <em>JUN</em> sequences, respectively. Each read consists of a UMI sequence, followed by a constant sequence and the variable region.</li>
<li>
<code>cisInput_{1,2}.fastq.gz</code>, <code>cisOutput_{1,2}.fastq.gz</code> - data from CIS experiment in <span class="citation">[@Diss2018]</span>. The forward and reverse reads both correspond to the mutated <em>FOS</em> sequence. Each read consists of a UMI sequence, followed by a constant sequence and the variable region.</li>
<li>
<code>GSE102901_cis_se.rds</code> - a <code>SummarizedExperiment</code> object obtained by processing the full CIS data from <span class="citation">[@Diss2018]</span>.</li>
<li>
<code>leujunt0_{1,2}.fastq.gz</code> - unpublished data; the forward read corresponds to an unmutated sequence of one of 46 leucine zipper sequences, and the reverse read corresponds to mutated <em>JUN</em> sequences. Each read contains a (frame-shifted) primer sequence, followed by the variable region.</li>
</ul>
</div>
<div class="section level2">
<h2 id="processing-deep-mutational-scanning-data">Processing Deep Mutational Scanning data<a class="anchor" aria-label="anchor" href="#processing-deep-mutational-scanning-data"></a>
</h2>
<div class="section level3">
<h3 id="read-composition-specification">Read composition specification<a class="anchor" aria-label="anchor" href="#read-composition-specification"></a>
</h3>
<p>The main function for processing the Deep Mutational Scanning FASTQ files is <code><a href="../reference/digestFastqs.html">digestFastqs()</a></code>. This function requires the specification of paths to compressed FASTQ file(s) and the composition of the reads in these files. The composition is specified by the user, and is given in the form of a character string indicating the parts constituting the respective read, and an integer vector specifying the lengths of the individual parts. The permitted components are:</p>
<ul>
<li>
<code>S</code> (skip) - nucleotide(s) to skip</li>
<li>
<code>U</code> (UMI) - UMI sequence</li>
<li>
<code>C</code> (constant) - constant sequence, can be used e.g. to estimate the sequencing error rate if available</li>
<li>
<code>V</code> (variable) - nucleotides corresponding to the variable region</li>
<li>
<code>P</code> (primer) - a primer sequence, which may appear in a nonspecified position in the read but whose position defines the start or end position of other components.</li>
</ul>
<p>A read can have several components of the same type (e.g., skipped nucleotides both in the beginning or the end, or two variable regions separated by a primer). In such cases, <code>mutscan</code> will concatenate all components of the same type in the analysis.</p>
<p>As an example, a read with the following composition:</p>
<p>[1 skipped nt] - [10 nt UMI] - [18 nt constant sequence] - [96 nt variable region]</p>
<p>would be specified to <code><a href="../reference/digestFastqs.html">digestFastqs()</a></code> by an elements string <code>"SUCV"</code>, and an element length vector <code>c(1, 10, 18, 96)</code>.</p>
<p>The package can also accommodate designs with primer sequences. In this situation, the primer acts as an ‘anchor’, and the read composition before and after the primer is specified. For example, a read with the following composition:</p>
<p>[unknown sequence] - [10 nt primer] - [variable region, constituting the remainder of the read]</p>
<p>would be represented by an elements string <code>"SPV"</code>, and an element length vector <code>c(-1, 10, -1)</code>, where the -1 indicates that the corresponding read part consists of the remaining part of the read, not accounted for by any of the other specified parts. In addition, the sequence of the primer must be specified, and any read where the primer is not present will be discarded.</p>
<p>The forward and reverse reads can have different compositions. The user can also specify whether the variable parts of the forward and/or reverse reads should be reverse complemented before being processed, and whether the variable regions of the forward and reverse reads should be merged into a single consensus sequence.</p>
</div>
<div class="section level3">
<h3 id="filtering">Filtering<a class="anchor" aria-label="anchor" href="#filtering"></a>
</h3>
<p>In addition to reading the FASTQ files, the <code><a href="../reference/digestFastqs.html">digestFastqs()</a></code> function will perform a series of filtering steps, in the following order:</p>
<ul>
<li>Search for perfect matches to forward/reverse adapter sequences, filter out the read pair if a match is found in either the forward or reverse read.</li>
<li>Search for perfect matches to forward/reverse primer sequences, filter out the read pair if not both are found.</li>
<li>Filter out reads whose length is not compatible with the indicated composition.</li>
<li>If forward and reverse variable regions should be merged, filter out read pairs where no valid overlap could be found.</li>
<li>Filter out read pair if the average quality in the variable region is below <code>avePhredMin</code> in either the forward or reverse read (or the consensus sequence if they are merged).</li>
<li>Filter out read pair if the number of Ns in the variable region exceeds <code>variableNMax</code>.</li>
<li>Filter out read pair if the number of Ns in the combined forward and reverse UMI sequence exceeds <code>umiNMax</code>.</li>
<li>Filter out read pair if any mutated base has a quality below <code>mutatedPhredMin</code>.</li>
<li>Filter out read pair if the number of mutated codons exceeds <code>nbrMutatedCodonsMax</code> or the number of mutated nucleotides exceeds <code>nbrMutatedBasesMax</code>.</li>
<li>Filter out read pair of there are multiple equally close best hits among the provided wild type sequences.</li>
<li>Filter out read pair if any of the mutated codons match any of the codons encoded by <code>forbiddenMutatedCodons</code>.</li>
<li>Filter out read pair if there are too many mutations in the constant sequence, or if there are multiple equally close best hits among the provided constant sequences.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="processing-trans-data">Processing TRANS data<a class="anchor" aria-label="anchor" href="#processing-trans-data"></a>
</h2>
<p>Here, we illustrate the processing of the provided TRANS experiment example data. We filter out reads with any adapter match, an average phred quality below 20, any Ns in the UMI or variable sequence, more than one mutated codon, or a mutated codon ending with A or T (represented by the <code>NNW</code> value of the <code>forbiddenMutatedCodonsForward</code> and <code>forbiddenMutatedCodonsReverse</code> arguments).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">transInput</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/digestFastqs.html">digestFastqs</a></span><span class="op">(</span>
    fastqForward <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"transInput_1.fastq.gz"</span><span class="op">)</span>,
    fastqReverse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"transInput_2.fastq.gz"</span><span class="op">)</span>,
    mergeForwardReverse <span class="op">=</span> <span class="cn">FALSE</span>, 
    revComplForward <span class="op">=</span> <span class="cn">FALSE</span>, revComplReverse <span class="op">=</span> <span class="cn">FALSE</span>,
    adapterForward <span class="op">=</span> <span class="st">"GGAAGAGCACACGTC"</span>, 
    adapterReverse <span class="op">=</span> <span class="st">"GGAAGAGCGTCGTGT"</span>, 
    elementsForward <span class="op">=</span> <span class="st">"SUCV"</span>, elementsReverse <span class="op">=</span> <span class="st">"SUCV"</span>,
    elementLengthsForward <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10</span>, <span class="fl">18</span>, <span class="fl">96</span><span class="op">)</span>,
    elementLengthsReverse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">8</span>, <span class="fl">20</span>, <span class="fl">96</span><span class="op">)</span>,
    constantForward <span class="op">=</span> <span class="st">"AACCGGAGGAGGGAGCTG"</span>,
    constantReverse <span class="op">=</span> <span class="st">"GAAAAAGGAAGCTGGAGAGA"</span>, 
    avePhredMinForward <span class="op">=</span> <span class="fl">20</span>, avePhredMinReverse <span class="op">=</span> <span class="fl">20</span>,
    variableNMaxForward <span class="op">=</span> <span class="fl">0</span>, variableNMaxReverse <span class="op">=</span> <span class="fl">0</span>, umiNMax <span class="op">=</span> <span class="fl">0</span>,
    wildTypeForward <span class="op">=</span> <span class="st">"ACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA"</span>, 
    wildTypeReverse <span class="op">=</span> <span class="st">"ATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTT"</span>, 
    nbrMutatedCodonsMaxForward <span class="op">=</span> <span class="fl">1</span>, nbrMutatedCodonsMaxReverse <span class="op">=</span> <span class="fl">1</span>,
    forbiddenMutatedCodonsForward <span class="op">=</span> <span class="st">"NNW"</span>, forbiddenMutatedCodonsReverse <span class="op">=</span> <span class="st">"NNW"</span>,
    mutNameDelimiter <span class="op">=</span> <span class="st">"."</span>, 
    constantMaxDistForward <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,
    constantMaxDistReverse <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,
    verbose <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
<p>The <code><a href="../reference/digestFastqs.html">digestFastqs()</a></code> function returns a list with four elements. The <code>parameters</code> list records all parameter values used during the processing, as well as the package version and time of processing.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">transInput</span><span class="op">$</span><span class="va">parameters</span>
<span class="co">#&gt; $fastqForward</span>
<span class="co">#&gt; [1] "/Users/runner/work/_temp/Library/mutscan/extdata/transInput_1.fastq.gz"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $fastqReverse</span>
<span class="co">#&gt; [1] "/Users/runner/work/_temp/Library/mutscan/extdata/transInput_2.fastq.gz"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $mergeForwardReverse</span>
<span class="co">#&gt; [1] FALSE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $minOverlap</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $maxOverlap</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $minMergedLength</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $maxMergedLength</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $maxFracMismatchOverlap</span>
<span class="co">#&gt; [1] 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $greedyOverlap</span>
<span class="co">#&gt; [1] TRUE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $revComplForward</span>
<span class="co">#&gt; [1] FALSE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $revComplReverse</span>
<span class="co">#&gt; [1] FALSE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $elementsForward</span>
<span class="co">#&gt; [1] "SUCV"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $elementLengthsForward</span>
<span class="co">#&gt; [1]  1 10 18 96</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $elementsReverse</span>
<span class="co">#&gt; [1] "SUCV"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $elementLengthsReverse</span>
<span class="co">#&gt; [1]  1  8 20 96</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $adapterForward</span>
<span class="co">#&gt; [1] "GGAAGAGCACACGTC"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $adapterReverse</span>
<span class="co">#&gt; [1] "GGAAGAGCGTCGTGT"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $primerForward</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $primerReverse</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $wildTypeForward</span>
<span class="co">#&gt;                                                                                                  f </span>
<span class="co">#&gt; "ACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA" </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $wildTypeReverse</span>
<span class="co">#&gt;                                                                                                  r </span>
<span class="co">#&gt; "ATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTT" </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $constantForward</span>
<span class="co">#&gt; [1] "AACCGGAGGAGGGAGCTG"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $constantReverse</span>
<span class="co">#&gt; [1] "GAAAAAGGAAGCTGGAGAGA"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $avePhredMinForward</span>
<span class="co">#&gt; [1] 20</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $avePhredMinReverse</span>
<span class="co">#&gt; [1] 20</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $variableNMaxForward</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $variableNMaxReverse</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $umiNMax</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $nbrMutatedCodonsMaxForward</span>
<span class="co">#&gt; [1] 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $nbrMutatedCodonsMaxReverse</span>
<span class="co">#&gt; [1] 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $nbrMutatedBasesMaxForward</span>
<span class="co">#&gt; [1] -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $nbrMutatedBasesMaxReverse</span>
<span class="co">#&gt; [1] -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $forbiddenMutatedCodonsForward</span>
<span class="co">#&gt;  [1] "AAA" "AAT" "ACA" "ACT" "AGA" "AGT" "ATA" "ATT" "CAA" "CAT" "CCA" "CCT"</span>
<span class="co">#&gt; [13] "CGA" "CGT" "CTA" "CTT" "GAA" "GAT" "GCA" "GCT" "GGA" "GGT" "GTA" "GTT"</span>
<span class="co">#&gt; [25] "TAA" "TAT" "TCA" "TCT" "TGA" "TGT" "TTA" "TTT"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $forbiddenMutatedCodonsReverse</span>
<span class="co">#&gt;  [1] "AAA" "AAT" "ACA" "ACT" "AGA" "AGT" "ATA" "ATT" "CAA" "CAT" "CCA" "CCT"</span>
<span class="co">#&gt; [13] "CGA" "CGT" "CTA" "CTT" "GAA" "GAT" "GCA" "GCT" "GGA" "GGT" "GTA" "GTT"</span>
<span class="co">#&gt; [25] "TAA" "TAT" "TCA" "TCT" "TGA" "TGT" "TTA" "TTT"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $useTreeWTmatch</span>
<span class="co">#&gt; [1] FALSE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $mutatedPhredMinForward</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $mutatedPhredMinReverse</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $mutNameDelimiter</span>
<span class="co">#&gt; [1] "."</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $constantMaxDistForward</span>
<span class="co">#&gt; [1] -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $constantMaxDistReverse</span>
<span class="co">#&gt; [1] -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $variableCollapseMaxDist</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $variableCollapseMinReads</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $variableCollapseMinRatio</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $umiCollapseMaxDist</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $filteredReadsFastqForward</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $filteredReadsFastqReverse</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $maxNReads</span>
<span class="co">#&gt; [1] -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $nThreads</span>
<span class="co">#&gt; [1] 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $chunkSize</span>
<span class="co">#&gt; [1] 100000</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $processingInfo</span>
<span class="co">#&gt; [1] "Processed by mutscan v0.2.26 on 2022-01-21 14:31:53"</span></code></pre></div>
<p>The <code>filterSummary</code> data.frame contains a summary of the number of reads filtered out in each step. Note that for any given filtering step, only the reads retained by the previous filters are considered. The numbers in the filter column names indicate the order of the filters.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">transInput</span><span class="op">$</span><span class="va">filterSummary</span>
<span class="co">#&gt;   nbrTotal f1_nbrAdapter f2_nbrNoPrimer f3_nbrReadWrongLength</span>
<span class="co">#&gt; 1     1000           314              0                     0</span>
<span class="co">#&gt;   f4_nbrNoValidOverlap f5_nbrAvgVarQualTooLow f6_nbrTooManyNinVar</span>
<span class="co">#&gt; 1                    0                      7                   0</span>
<span class="co">#&gt;   f7_nbrTooManyNinUMI f8_nbrTooManyBestWTHits f9_nbrMutQualTooLow</span>
<span class="co">#&gt; 1                   0                       0                   0</span>
<span class="co">#&gt;   f10a_nbrTooManyMutCodons f10b_nbrTooManyMutBases f11_nbrForbiddenCodons</span>
<span class="co">#&gt; 1                      392                       0                      8</span>
<span class="co">#&gt;   f12_nbrTooManyMutConstant f13_nbrTooManyBestConstantHits nbrRetained</span>
<span class="co">#&gt; 1                         0                              0         279</span></code></pre></div>
<p>The <code>summaryTable</code> provides the number of reads and unique UMIs observed for each variable sequence pair. In addition, it contains a column <code>mutantName</code>, which provides a shorthand notation for each mutant. The format of the values in this column is a combination of <code>WTS.xx.NNN</code> (separated by <code>_</code>), where <code>WTS</code> provides the name of the closest matching wild type sequence (if only one, unnamed wild type sequence is provided, the name will be <code>f/r</code> depending on if it corresponds to the forward/reverse read, respectively). The <code>xx</code> part indicates the mutated codon or nucleotide number, and <code>NNN</code> gives the observed sequence for the mutated codon or nucleotide. A sequence without mutations is named <code>WTS.0.WT</code>, where, again, <code>WTS</code> is the name of the matching wild type sequence.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">transInput</span><span class="op">$</span><span class="va">summaryTable</span><span class="op">)</span>
<span class="co">#&gt;        mutantName</span>
<span class="co">#&gt; 1 f.0.WT_r.13.CCC</span>
<span class="co">#&gt; 2 f.0.WT_r.13.CTC</span>
<span class="co">#&gt; 3  f.0.WT_r.2.TCC</span>
<span class="co">#&gt; 4 f.0.WT_r.20.ACC</span>
<span class="co">#&gt; 5 f.0.WT_r.30.AGG</span>
<span class="co">#&gt; 6 f.0.WT_r.30.GGG</span>
<span class="co">#&gt;                                                                                                                                                                                            sequence</span>
<span class="co">#&gt; 1 ACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA_ATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAACCCCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTT</span>
<span class="co">#&gt; 2 ACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA_ATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAACTCCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTT</span>
<span class="co">#&gt; 3 ACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA_ATCTCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTT</span>
<span class="co">#&gt; 4 ACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA_ATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGACCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTT</span>
<span class="co">#&gt; 5 ACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA_ATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGAGGCAGCTT</span>
<span class="co">#&gt; 6 ACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA_ATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGGGCAGCTT</span>
<span class="co">#&gt;   nbrMutBases nbrMutCodons nbrReads maxNbrReads nbrUmis</span>
<span class="co">#&gt; 1           2            1        1           1       1</span>
<span class="co">#&gt; 2           3            1        1           1       1</span>
<span class="co">#&gt; 3           1            1        1           1       1</span>
<span class="co">#&gt; 4           1            1        1           1       1</span>
<span class="co">#&gt; 5           3            1        1           1       1</span>
<span class="co">#&gt; 6           2            1        1           1       1</span></code></pre></div>
<p>Finally, the <code>errorStatistics</code> data.frame lists the number of matching and mismatching bases in the constant sequences, stratified by the phred quality score (from 0 to 99).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">transInput</span><span class="op">$</span><span class="va">errorStatistics</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">transInput</span><span class="op">$</span><span class="va">errorStatistics</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">!=</span> <span class="fl">0</span>, <span class="op">]</span>
<span class="co">#&gt;    PhredQuality nbrMatchForward nbrMismatchForward nbrMatchReverse</span>
<span class="co">#&gt; 15           14             160                 11             204</span>
<span class="co">#&gt; 23           22              52                  0              14</span>
<span class="co">#&gt; 28           27             302                  4             474</span>
<span class="co">#&gt; 34           33             472                  0             462</span>
<span class="co">#&gt; 38           37            4020                  1            4405</span>
<span class="co">#&gt;    nbrMismatchReverse</span>
<span class="co">#&gt; 15                 17</span>
<span class="co">#&gt; 23                  0</span>
<span class="co">#&gt; 28                  3</span>
<span class="co">#&gt; 34                  1</span>
<span class="co">#&gt; 38                  0</span></code></pre></div>
<p>The <code>errorStatistics</code> output can be used to estimate the sequencing error rate:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">(</span><span class="va">propErrorsConstantF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">transInput</span><span class="op">$</span><span class="va">errorStatistics</span><span class="op">$</span><span class="va">nbrMismatchForward</span><span class="op">)</span> <span class="op">/</span>
   <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">transInput</span><span class="op">$</span><span class="va">parameters</span><span class="op">$</span><span class="va">constantForward</span><span class="op">)</span> <span class="op">*</span> <span class="va">transInput</span><span class="op">$</span><span class="va">filterSummary</span><span class="op">$</span><span class="va">nbrRetained</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.003185982</span>
<span class="op">(</span><span class="va">propErrorsConstantR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">transInput</span><span class="op">$</span><span class="va">errorStatistics</span><span class="op">$</span><span class="va">nbrMismatchReverse</span><span class="op">)</span> <span class="op">/</span>
   <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">transInput</span><span class="op">$</span><span class="va">parameters</span><span class="op">$</span><span class="va">constantReverse</span><span class="op">)</span> <span class="op">*</span> <span class="va">transInput</span><span class="op">$</span><span class="va">filterSummary</span><span class="op">$</span><span class="va">nbrRetained</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.003763441</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="processing-cis-data">Processing CIS data<a class="anchor" aria-label="anchor" href="#processing-cis-data"></a>
</h2>
<p>Here, we illustrate the processing of the provided CIS experiment example data. As for the TRANS data, we filter out reads with any adapter match, an average phred quality below 20, any Ns in the UMI or variable sequence, more than one mutated codon, or a mutated codon ending with A or T (represented by the <code>NNW</code> value of the <code>forbiddenMutatedCodonsForward</code> argument). In this case, the forward and reverse variable sequences will be merged into one variable sequence (since <code>mergeForwardReverse</code> is <code>TRUE</code>) and thus only one wild type sequence is provided (the <code>wildTypeReverse</code> argument will be ignored if specified).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cisInput</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/digestFastqs.html">digestFastqs</a></span><span class="op">(</span>
    fastqForward <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"cisInput_1.fastq.gz"</span><span class="op">)</span>,
    fastqReverse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"cisInput_2.fastq.gz"</span><span class="op">)</span>,
    mergeForwardReverse <span class="op">=</span> <span class="cn">TRUE</span>, 
    minOverlap <span class="op">=</span> <span class="fl">0</span>, maxOverlap <span class="op">=</span> <span class="fl">0</span>, 
    maxFracMismatchOverlap <span class="op">=</span> <span class="fl">1</span>, greedyOverlap <span class="op">=</span> <span class="cn">TRUE</span>, 
    revComplForward <span class="op">=</span> <span class="cn">FALSE</span>, revComplReverse <span class="op">=</span> <span class="cn">TRUE</span>,
    adapterForward <span class="op">=</span> <span class="st">"GGAAGAGCACACGTC"</span>, 
    adapterReverse <span class="op">=</span> <span class="st">"GGAAGAGCGTCGTGT"</span>, 
    elementsForward <span class="op">=</span> <span class="st">"SUCV"</span>, elementsReverse <span class="op">=</span> <span class="st">"SUCVS"</span>,
    elementLengthsForward <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10</span>, <span class="fl">18</span>, <span class="fl">96</span><span class="op">)</span>,
    elementLengthsReverse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">7</span>, <span class="fl">17</span>, <span class="fl">96</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span>,
    constantForward <span class="op">=</span> <span class="st">"AACCGGAGGAGGGAGCTG"</span>,
    constantReverse <span class="op">=</span> <span class="st">"GAGTTCATCCTGGCAGC"</span>, 
    primerForward <span class="op">=</span> <span class="st">""</span>, primerReverse <span class="op">=</span> <span class="st">""</span>,
    avePhredMinForward <span class="op">=</span> <span class="fl">20</span>, avePhredMinReverse <span class="op">=</span> <span class="fl">20</span>,
    variableNMaxForward <span class="op">=</span> <span class="fl">0</span>, variableNMaxReverse <span class="op">=</span> <span class="fl">0</span>, umiNMax <span class="op">=</span> <span class="fl">0</span>,
    wildTypeForward <span class="op">=</span> <span class="st">"ACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA"</span>, 
    wildTypeReverse <span class="op">=</span> <span class="st">""</span>, 
    nbrMutatedCodonsMaxForward <span class="op">=</span> <span class="fl">1</span>, nbrMutatedCodonsMaxReverse <span class="op">=</span> <span class="fl">1</span>, 
    forbiddenMutatedCodonsForward <span class="op">=</span> <span class="st">"NNW"</span>, forbiddenMutatedCodonsReverse <span class="op">=</span> <span class="st">"NNW"</span>,
    mutNameDelimiter <span class="op">=</span> <span class="st">"."</span>, 
    constantMaxDistForward <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,
    constantMaxDistReverse <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,
    verbose <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>
<span class="co">#&gt; done enumerating forbidden codons (32)</span>
<span class="co">#&gt; done enumerating forbidden codons (32)</span>
<span class="co">#&gt; start reading sequences for file pair 1 of 1...</span>
<span class="co">#&gt;     1000 read pairs processed (16.7% retained)</span>
<span class="co">#&gt; done reading sequences</span>
<span class="co">#&gt; retained 67 unique features</span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cisInput</span><span class="op">$</span><span class="va">parameters</span>
<span class="co">#&gt; $fastqForward</span>
<span class="co">#&gt; [1] "/Users/runner/work/_temp/Library/mutscan/extdata/cisInput_1.fastq.gz"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $fastqReverse</span>
<span class="co">#&gt; [1] "/Users/runner/work/_temp/Library/mutscan/extdata/cisInput_2.fastq.gz"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $mergeForwardReverse</span>
<span class="co">#&gt; [1] TRUE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $minOverlap</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $maxOverlap</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $minMergedLength</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $maxMergedLength</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $maxFracMismatchOverlap</span>
<span class="co">#&gt; [1] 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $greedyOverlap</span>
<span class="co">#&gt; [1] TRUE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $revComplForward</span>
<span class="co">#&gt; [1] FALSE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $revComplReverse</span>
<span class="co">#&gt; [1] TRUE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $elementsForward</span>
<span class="co">#&gt; [1] "SUCV"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $elementLengthsForward</span>
<span class="co">#&gt; [1]  1 10 18 96</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $elementsReverse</span>
<span class="co">#&gt; [1] "SUCVS"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $elementLengthsReverse</span>
<span class="co">#&gt; [1]  1  7 17 96 -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $adapterForward</span>
<span class="co">#&gt; [1] "GGAAGAGCACACGTC"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $adapterReverse</span>
<span class="co">#&gt; [1] "GGAAGAGCGTCGTGT"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $primerForward</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $primerReverse</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $wildTypeForward</span>
<span class="co">#&gt;                                                                                                  f </span>
<span class="co">#&gt; "ACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA" </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $wildTypeReverse</span>
<span class="co">#&gt;  r </span>
<span class="co">#&gt; "" </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $constantForward</span>
<span class="co">#&gt; [1] "AACCGGAGGAGGGAGCTG"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $constantReverse</span>
<span class="co">#&gt; [1] "GAGTTCATCCTGGCAGC"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $avePhredMinForward</span>
<span class="co">#&gt; [1] 20</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $avePhredMinReverse</span>
<span class="co">#&gt; [1] 20</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $variableNMaxForward</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $variableNMaxReverse</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $umiNMax</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $nbrMutatedCodonsMaxForward</span>
<span class="co">#&gt; [1] 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $nbrMutatedCodonsMaxReverse</span>
<span class="co">#&gt; [1] 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $nbrMutatedBasesMaxForward</span>
<span class="co">#&gt; [1] -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $nbrMutatedBasesMaxReverse</span>
<span class="co">#&gt; [1] -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $forbiddenMutatedCodonsForward</span>
<span class="co">#&gt;  [1] "AAA" "AAT" "ACA" "ACT" "AGA" "AGT" "ATA" "ATT" "CAA" "CAT" "CCA" "CCT"</span>
<span class="co">#&gt; [13] "CGA" "CGT" "CTA" "CTT" "GAA" "GAT" "GCA" "GCT" "GGA" "GGT" "GTA" "GTT"</span>
<span class="co">#&gt; [25] "TAA" "TAT" "TCA" "TCT" "TGA" "TGT" "TTA" "TTT"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $forbiddenMutatedCodonsReverse</span>
<span class="co">#&gt;  [1] "AAA" "AAT" "ACA" "ACT" "AGA" "AGT" "ATA" "ATT" "CAA" "CAT" "CCA" "CCT"</span>
<span class="co">#&gt; [13] "CGA" "CGT" "CTA" "CTT" "GAA" "GAT" "GCA" "GCT" "GGA" "GGT" "GTA" "GTT"</span>
<span class="co">#&gt; [25] "TAA" "TAT" "TCA" "TCT" "TGA" "TGT" "TTA" "TTT"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $useTreeWTmatch</span>
<span class="co">#&gt; [1] FALSE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $mutatedPhredMinForward</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $mutatedPhredMinReverse</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $mutNameDelimiter</span>
<span class="co">#&gt; [1] "."</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $constantMaxDistForward</span>
<span class="co">#&gt; [1] -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $constantMaxDistReverse</span>
<span class="co">#&gt; [1] -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $variableCollapseMaxDist</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $variableCollapseMinReads</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $variableCollapseMinRatio</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $umiCollapseMaxDist</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $filteredReadsFastqForward</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $filteredReadsFastqReverse</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $maxNReads</span>
<span class="co">#&gt; [1] -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $nThreads</span>
<span class="co">#&gt; [1] 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $chunkSize</span>
<span class="co">#&gt; [1] 100000</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $processingInfo</span>
<span class="co">#&gt; [1] "Processed by mutscan v0.2.26 on 2022-01-21 14:31:54"</span>
<span class="va">cisInput</span><span class="op">$</span><span class="va">filterSummary</span>
<span class="co">#&gt;   nbrTotal f1_nbrAdapter f2_nbrNoPrimer f3_nbrReadWrongLength</span>
<span class="co">#&gt; 1     1000           126              0                     0</span>
<span class="co">#&gt;   f4_nbrNoValidOverlap f5_nbrAvgVarQualTooLow f6_nbrTooManyNinVar</span>
<span class="co">#&gt; 1                    0                      0                  44</span>
<span class="co">#&gt;   f7_nbrTooManyNinUMI f8_nbrTooManyBestWTHits f9_nbrMutQualTooLow</span>
<span class="co">#&gt; 1                   0                       0                   0</span>
<span class="co">#&gt;   f10a_nbrTooManyMutCodons f10b_nbrTooManyMutBases f11_nbrForbiddenCodons</span>
<span class="co">#&gt; 1                      581                       0                     82</span>
<span class="co">#&gt;   f12_nbrTooManyMutConstant f13_nbrTooManyBestConstantHits nbrRetained</span>
<span class="co">#&gt; 1                         0                              0         167</span>
<span class="va">cisInput</span><span class="op">$</span><span class="va">errorStatistics</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">cisInput</span><span class="op">$</span><span class="va">errorStatistics</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">!=</span> <span class="fl">0</span>, <span class="op">]</span>
<span class="co">#&gt;    PhredQuality nbrMatchForward nbrMismatchForward nbrMatchReverse</span>
<span class="co">#&gt; 3             2               0                  0               0</span>
<span class="co">#&gt; 15           14              51                  1              41</span>
<span class="co">#&gt; 23           22              19                  0               0</span>
<span class="co">#&gt; 28           27              79                  1              56</span>
<span class="co">#&gt; 34           33             163                  0              96</span>
<span class="co">#&gt; 38           37            2690                  2            2620</span>
<span class="co">#&gt;    nbrMismatchReverse</span>
<span class="co">#&gt; 3                   4</span>
<span class="co">#&gt; 15                  7</span>
<span class="co">#&gt; 23                  0</span>
<span class="co">#&gt; 28                  0</span>
<span class="co">#&gt; 34                  1</span>
<span class="co">#&gt; 38                 14</span></code></pre></div>
<p>The <code>summaryTable</code> now provides the number of reads and unique UMIs observed for each variable sequence, and all values in the <code>mutNames</code> column will start with <code>f</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cisInput</span><span class="op">$</span><span class="va">summaryTable</span><span class="op">)</span>
<span class="co">#&gt;   mutantName</span>
<span class="co">#&gt; 1     f.0.WT</span>
<span class="co">#&gt; 2    f.1.ACC</span>
<span class="co">#&gt; 3    f.1.ACG</span>
<span class="co">#&gt; 4   f.11.CTG</span>
<span class="co">#&gt; 5   f.13.GAC</span>
<span class="co">#&gt; 6   f.13.GAG</span>
<span class="co">#&gt;                                                                                           sequence</span>
<span class="co">#&gt; 1 ACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA</span>
<span class="co">#&gt; 2 ACCGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA</span>
<span class="co">#&gt; 3 ACGGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA</span>
<span class="co">#&gt; 4 ACTGATACACTCCAAGCGGAGACAGACCAACTGGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA</span>
<span class="co">#&gt; 5 ACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGACGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA</span>
<span class="co">#&gt; 6 ACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGAGGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA</span>
<span class="co">#&gt;   nbrMutBases nbrMutCodons nbrReads maxNbrReads nbrUmis</span>
<span class="co">#&gt; 1           0            0       77          77      77</span>
<span class="co">#&gt; 2           1            1        2           2       2</span>
<span class="co">#&gt; 3           1            1        1           1       1</span>
<span class="co">#&gt; 4           1            1        1           1       1</span>
<span class="co">#&gt; 5           1            1        1           1       1</span>
<span class="co">#&gt; 6           1            1        1           1       1</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="processing-trans-data-with-primers">Processing TRANS data with primers<a class="anchor" aria-label="anchor" href="#processing-trans-data-with-primers"></a>
</h2>
<p>Here, we illustrate the processing of the provided example data, where the first read corresponds to one of 46 leucine zipper sequences, and the second read is a mutated <em>JUN</em> sequence. We first need to define the possible sequences for the forward read. If multiple wild type sequences are provided like here, <code>mutscan</code> will match each read against all of them and find the most similar one in each case.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">leu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>
    ATF2 <span class="op">=</span> <span class="st">"GATCCTGATGAAAAAAGGAGAAAGTTTTTAGAGCGAAATAGAGCAGCAGCTTCAAGATGCCGACAAAAAAGGAAAGTCTGGGTTCAGTCTTTAGAGAAGAAAGCTGAAGACTTGAGTTCATTAAATGGTCAGCTGCAGAGTGAAGTCACCCTGCTGAGAAATGAAGTGGCACAGCTGAAACAGCTTCTTCTGGCT"</span>,
    ATF7 <span class="op">=</span> <span class="st">"GATCCAGATGAGCGACGGCAGCGCTTTCTGGAGCGCAACCGGGCTGCAGCCTCCCGCTGCCGCCAAAAGCGAAAGCTGTGGGTGTCCTCCCTAGAGAAGAAGGCCGAAGAACTCACTTCTCAGAACATTCAGCTGAGTAATGAAGTCACATTACTACGCAATGAGGTGGCCCAGTTGAAACAGCTACTGTTAGCT"</span>,
    CREB5 <span class="op">=</span> <span class="st">"GATCCGGACGAGAGGCGGCGGAAATTTCTGGAACGGAACCGGGCAGCTGCCACCCGCTGCAGACAGAAGAGGAAGGTCTGGGTGATGTCATTGGAAAAGAAAGCAGAAGAACTCACCCAGACAAACATGCAGCTTCAGAATGAAGTGTCTATGTTGAAAAATGAGGTGGCCCAGCTGAAACAGTTGTTGTTAACA"</span>,
    ATF3 <span class="op">=</span> <span class="st">"GAAGAAGATGAAAGGAAAAAGAGGCGACGAGAAAGAAATAAGATTGCAGCTGCAAAGTGCCGAAACAAGAAGAAGGAGAAGACGGAGTGCCTGCAGAAAGAGTCGGAGAAGCTGGAAAGTGTGAATGCTGAACTGAAGGCTCAGATTGAGGAGCTCAAGAACGAGAAGCAGCATTTGATATACATGCTCAACCTT"</span>,
    JDP2 <span class="op">=</span> <span class="st">"GAGGAAGAGGAGCGAAGGAAAAGGCGCCGGGAGAAGAACAAAGTCGCAGCAGCCCGATGCCGGAACAAGAAGAAGGAGCGCACGGAGTTTCTGCAGCGGGAATCCGAGCGGCTGGAACTCATGAACGCAGAGCTGAAGACCCAGATTGAGGAGCTGAAGCAGGAGCGGCAGCAGCTCATCCTGATGCTGAACCGA"</span>,
    ATF4 <span class="op">=</span> <span class="st">"GAGAAACTGGATAAGAAGCTGAAAAAAATGGAGCAAAACAAGACAGCAGCCACTAGGTACCGCCAGAAGAAGAGGGCGGAGCAGGAGGCTCTTACTGGTGAGTGCAAAGAGCTGGAAAAGAAGAACGAGGCTCTAAAAGAGAGGGCGGATTCCCTGGCCAAGGAGATCCAGTACCTGAAAGATTTGATAGAAGAG"</span>,
    ATF5 <span class="op">=</span> <span class="st">"ACCCGAGGGGACCGCAAGCAAAAGAAGAGAGACCAGAACAAGTCGGCGGCTCTGAGGTACCGCCAGCGGAAGCGGGCAGAGGGTGAGGCCCTGGAGGGCGAGTGCCAGGGGCTGGAGGCACGGAATCGCGAGCTGAAGGAACGGGCAGAGTCCGTGGAGCGCGAGATCCAGTACGTCAAGGACCTGCTCATCGAG"</span>,
    CREBZF <span class="op">=</span> <span class="st">"AGTCCCCGGAAGGCGGCGGCGGCCGCTGCCCGCCTTAATCGACTGAAGAAGAAGGAGTACGTGATGGGGCTGGAGAGTCGAGTCCGGGGTCTGGCAGCCGAGAACCAGGAGCTGCGGGCCGAGAATCGGGAGCTGGGCAAACGCGTACAGGCACTGCAGGAGGAGAGTCGCTACCTACGGGCAGTCTTAGCCAAC"</span>,
    BATF2 <span class="op">=</span> <span class="st">"CCCAAGGAGCAACAAAGGCAGCTGAAGAAGCAGAAGAACCGGGCAGCCGCCCAGCGAAGCCGGCAGAAGCACACAGACAAGGCAGACGCCCTGCACCAGCAGCACGAGTCTCTGGAAAAAGACAACCTCGCCCTGCGGAAGGAGATCCAGTCCCTGCAGGCCGAGCTGGCGTGGTGGAGCCGGACCCTGCACGTG"</span>,
    BATF3 <span class="op">=</span> <span class="st">"GAGGATGATGACAGGAAGGTCCGAAGGAGAGAAAAAAACCGAGTTGCTGCTCAGAGAAGTCGGAAGAAGCAGACCCAGAAGGCTGACAAGCTCCATGAGGAATATGAGAGCCTGGAGCAAGAAAACACCATGCTGCGGAGAGAGATCGGGAAGCTGACAGAGGAGCTGAAGCACCTGACAGAGGCACTGAAGGAG"</span>,
    CEBPE <span class="op">=</span> <span class="st">"AAAGATAGCCTTGAGTACCGGCTGAGGCGGGAGCGCAACAACATCGCCGTGCGCAAGAGCCGAGACAAGGCCAAGAGGCGCATTCTGGAGACGCAGCAGAAGGTGCTGGAGTACATGGCAGAGAACGAGCGCCTCCGCAGCCGCGTGGAGCAGCTCACCCAGGAGCTAGACACCCTCCGCAACCTCTTCCGCCAG"</span>,
    BACH1 <span class="op">=</span> <span class="st">"CTGGATTGTATCCATGATATTCGAAGAAGAAGTAAAAACAGAATTGCTGCACAGCGCTGTCGCAAGAGAAAACTTGACTGTATACAGAATCTTGAATCAGAAATTGAGAAGCTGCAAAGTGAAAAGGAGAGCTTGTTGAAGGAAAGAGATCACATTTTGTCAACTCTGGGTGAGACAAAGCAGAACCTAACTGGA"</span>,
    BACH2 <span class="op">=</span> <span class="st">"TTAGAGTTTATTCATGATGTCCGACGGCGCAGCAAGAACCGCATCGCGGCCCAGCGCTGCCGCAAAAGGAAACTGGACTGTATTCAGAATTTAGAATGTGAAATCCGCAAATTGGTGTGTGAGAAAGAGAAACTGTTGTCAGAGAGGAATCAACTGAAAGCATGCATGGGGGAACTGTTGGACAACTTCTCCTGC"</span>,
    NFE2L1 <span class="op">=</span> <span class="st">"CTGAGCCTCATCCGAGACATCCGGCGCCGGGGCAAGAACAAGATGGCGGCGCAGAACTGCCGCAAGCGCAAGCTGGACACCATCCTGAATCTGGAGCGTGATGTGGAGGACCTGCAGCGTGACAAAGCCCGGCTGCTGCGGGAGAAAGTGGAGTTCCTGCGCTCCCTGCGACAGATGAAGCAGAAGGTCCAGAGC"</span>,
    NFE2 <span class="op">=</span> <span class="st">"CTAGCGCTAGTCCGGGACATCCGACGACGGGGCAAAAACAAGGTGGCAGCCCAGAACTGCCGCAAGAGGAAGCTGGAAACCATTGTGCAGCTGGAGCGGGAGCTGGAGCGGCTGACCAATGAACGGGAGCGGCTTCTCAGGGCCCGCGGGGAGGCAGACCGGACCCTGGAGGTCATGCGCCAACAGCTGACAGAG"</span>,
    NFIL3 <span class="op">=</span> <span class="st">"AAGAAAGATGCTATGTATTGGGAAAAAAGGCGGAAAAATAATGAAGCTGCCAAAAGATCTCGTGAGAAGCGTCGACTGAATGACCTGGTTTTAGAGAACAAACTAATTGCACTGGGAGAAGAAAACGCCACTTTAAAAGCTGAGCTGCTTTCACTAAAATTAAAGTTTGGTTTAATTAGCTCCACAGCATATGCT"</span>,
    FOS <span class="op">=</span> <span class="st">"GAAGAAGAAGAGAAAAGGAGAATCCGAAGGGAAAGGAATAAGATGGCTGCAGCCAAATGCCGCAACCGGAGGAGGGAGCTGACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTAGAGTTCATCCTGGCAGCT"</span>,
    FOSB <span class="op">=</span> <span class="st">"GAGGAAGAGGAGAAGCGAAGGGTGCGCCGGGAACGAAATAAACTAGCAGCAGCTAAATGCAGGAACCGGCGGAGGGAGCTGACCGACCGACTCCAGGCGGAGACAGATCAGTTGGAGGAAGAAAAAGCAGAGCTGGAGTCGGAGATCGCCGAGCTCCAAAAGGAGAAGGAACGTCTGGAGTTTGTGCTGGTGGCC"</span>,
    FOSL1 <span class="op">=</span> <span class="st">"GAGGAAGAGGAGCGCCGCCGAGTAAGGCGCGAGCGGAACAAGCTGGCTGCGGCCAAGTGCAGGAACCGGAGGAAGGAACTGACCGACTTCCTGCAGGCGGAGACTGACAAACTGGAAGATGAGAAATCTGGGCTGCAGCGAGAGATTGAGGAGCTGCAGAAGCAGAAGGAGCGCCTAGAGCTGGTGCTGGAAGCC"</span>,
    FOSL2 <span class="op">=</span> <span class="st">"GAAGAGGAGGAGAAGCGTCGCATCCGGCGGGAGAGGAACAAGCTGGCTGCAGCCAAGTGCCGGAACCGACGCCGGGAGCTGACAGAGAAGCTGCAGGCGGAGACAGAGGAGCTGGAGGAGGAGAAGTCAGGCCTGCAGAAGGAGATTGCTGAGCTGCAGAAGGAGAAGGAGAAGCTGGAGTTCATGTTGGTGGCT"</span>,
    MAFB <span class="op">=</span> <span class="st">"GTGATCCGCCTGAAGCAGAAGCGGCGGACCCTGAAGAACCGGGGCTACGCCCAGTCTTGCAGGTATAAACGCGTCCAGCAGAAGCACCACCTGGAGAATGAGAAGACGCAGCTCATTCAGCAGGTGGAGCAGCTTAAGCAGGAGGTGTCCCGGCTGGCCCGCGAGAGAGACGCCTACAAGGTCAAGTGCGAGAAA"</span>,
    JUN <span class="op">=</span> <span class="st">"CAGGAGCGGATCAAGGCGGAGAGGAAGCGCATGAGGAACCGCATCGCTGCCTCCAAGTGCCGAAAAAGGAAGCTGGAGAGAATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTTAAACAGAAAGTCATGAAC"</span>,
    JUNB <span class="op">=</span> <span class="st">"CAAGAGCGCATCAAAGTGGAGCGCAAGCGGCTGCGGAACCGGCTGGCGGCCACCAAGTGCCGGAAGCGGAAGCTGGAGCGCATCGCGCGCCTGGAGGACAAGGTGAAGACGCTCAAGGCCGAGAACGCGGGGCTGTCGAGTACCGCCGGCCTCCTCCGGGAGCAGGTGGCCCAGCTCAAACAGAAGGTCATGACC"</span>,
    JUND <span class="op">=</span> <span class="st">"CAGGAGCGCATCAAGGCGGAGCGCAAGCGGCTGCGCAACCGCATCGCCGCCTCCAAGTGCCGCAAGCGCAAGCTGGAGCGCATCTCGCGCCTGGAAGAGAAAGTGAAGACCCTCAAGAGTCAGAACACGGAGCTGGCGTCCACGGCGAGCCTGCTGCGCGAGCAGGTGGCGCAGCTCAAGCAGAAAGTCCTCAGC"</span>,
    CREB3 <span class="op">=</span> <span class="st">"GAACAAATTCTGAAACGTGTGCGGAGGAAGATTCGAAATAAAAGATCTGCTCAAGAGAGCCGCAGGAAAAAGAAGGTGTATGTTGGGGGTTTAGAGAGCAGGGTCTTGAAATACACAGCCCAGAATATGGAGCTTCAGAACAAAGTACAGCTTCTGGAGGAACAGAATTTGTCCCTTCTAGATCAACTGAGGAAA"</span>,
    HLF <span class="op">=</span> <span class="st">"CTGAAGGATGACAAGTACTGGGCAAGGCGCAGAAAGAACAACATGGCAGCCAAGCGCTCCCGCGACGCCCGGAGGCTGAAAGAGAACCAGATCGCCATCCGGGCCTCGTTCCTGGAGAAGGAGAACTCGGCCCTCCGCCAGGAGGTGGCTGACTTGAGGAAGGAGCTGGGCAAATGCAAGAACATACTTGCCAAG"</span>,
    MAFG <span class="op">=</span> <span class="st">"ATCGTCCAGCTGAAGCAGCGCCGGCGCACGCTCAAGAACCGCGGCTACGCTGCCAGCTGCCGCGTGAAGCGGGTGACGCAGAAGGAGGAGCTGGAGAAGCAGAAGGCGGAGCTGCAGCAGGAGGTGGAGAAGCTGGCCTCAGAGAACGCCAGCATGAAGCTGGAGCTCGACGCGCTGCGCTCCAAGTACGAGGCG"</span>,
    MAFK <span class="op">=</span> <span class="st">"GTGACCCGCCTGAAGCAGCGTCGGCGCACACTCAAGAACCGCGGCTACGCGGCCAGCTGCCGCATCAAGCGGGTGACGCAGAAGGAGGAGCTGGAGCGGCAGCGCGTGGAGCTGCAGCAGGAGGTGGAGAAGCTGGCGCGTGAGAACAGCAGCATGCGGCTGGAGCTGGACGCCCTGCGCTCCAAGTACGAGGCG"</span>,
    XBP1 <span class="op">=</span> <span class="st">"AGCCCCGAGGAGAAGGCGCTGAGGAGGAAACTGAAAAACAGAGTAGCAGCTCAGACTGCCAGAGATCGAAAGAAGGCTCGAATGAGTGAGCTGGAACAGCAAGTGGTAGATTTAGAAGAAGAGAACCAAAAACTTTTGCTAGAAAATCAGCTTTTACGAGAGAAAACTCATGGCCTTGTAGTTGAGAACCAGGAG"</span>,
    ATF6 <span class="op">=</span> <span class="st">"ATTGCTGTGCTAAGGAGACAGCAACGTATGATAAAAAATCGAGAATCCGCTTGTCAGTCTCGCAAGAAGAAGAAAGAATATATGCTAGGGTTAGAGGCGAGATTAAAGGCTGCCCTCTCAGAAAACGAGCAACTGAAGAAAGAAAATGGAACACTGAAGCGGCAGCTGGATGAAGTTGTGTCAGAGAACCAGAGG"</span>,
    ATF6B <span class="op">=</span> <span class="st">"GCAAAGCTGCTGAAGCGGCAGCAGCGAATGATCAAGAACCGGGAGTCAGCCTGCCAGTCCCGGAGAAAGAAGAAAGAGTATCTGCAGGGACTGGAGGCTCGGCTGCAAGCAGTACTGGCTGACAACCAGCAGCTCCGCCGAGAGAATGCTGCCCTCCGGCGGCGGCTGGAGGCCCTGCTGGCTGAAAACAGCGAG"</span>,
    CEBPA <span class="op">=</span> <span class="st">"AAGAACAGCAACGAGTACCGGGTGCGGCGCGAGCGCAACAACATCGCGGTGCGCAAGAGCCGCGACAAGGCCAAGCAGCGCAACGTGGAGACGCAGCAGAAGGTGCTGGAGCTGACCAGTGACAATGACCGCCTGCGCAAGCGGGTGGAACAGCTGAGCCGCGAACTGGACACGCTGCGGGGCATCTTCCGCCAG"</span>,
    CEBPB <span class="op">=</span> <span class="st">"AAGCACAGCGACGAGTACAAGATCCGGCGCGAGCGCAACAACATCGCCGTGCGCAAGAGCCGCGACAAGGCCAAGATGCGCAACCTGGAGACGCAGCACAAGGTCCTGGAGCTCACGGCCGAGAACGAGCGGCTGCAGAAGAAGGTGGAGCAGCTGTCGCGCGAGCTCAGCACCCTGCGGAACTTGTTCAAGCAG"</span>,
    CEBPD <span class="op">=</span> <span class="st">"CGCGGCAGCCCCGAGTACCGGCAGCGGCGCGAGCGCAACAACATCGCCGTGCGCAAGAGCCGCGACAAGGCCAAGCGGCGCAACCAGGAGATGCAGCAGAAGTTGGTGGAGCTGTCGGCTGAGAACGAGAAGCTGCACCAGCGCGTGGAGCAGCTCACGCGGGACCTGGCCGGCCTCCGGCAGTTCTTCAAGCAG"</span>,
    CEBPG <span class="op">=</span> <span class="st">"CGAAACAGTGACGAGTATCGGCAACGCCGAGAGAGGAACAACATGGCTGTGAAAAAGAGCCGGTTGAAAAGCAAGCAGAAAGCACAAGACACACTGCAGAGAGTCAATCAGCTCAAAGAAGAGAATGAACGGTTGGAAGCAAAAATCAAATTGCTGACCAAGGAATTAAGTGTACTCAAAGATTTGTTTCTTGAG"</span>,
    CREB1 <span class="op">=</span> <span class="st">"GAAGCAGCACGAAAGAGAGAGGTCCGTCTAATGAAGAACAGGGAAGCAGCTCGAGAGTGTCGTAGAAAGAAGAAAGAATATGTGAAATGTTTAGAAAACAGAGTGGCAGTGCTTGAAAATCAAAACAAGACATTGATTGAGGAGCTAAAAGCACTTAAGGACCTTTACTGCCACAAATCAGAT"</span>,
    CREB3L1 <span class="op">=</span> <span class="st">"GAGAAGGCCTTGAAGAGAGTCCGGAGGAAAATCAAGAACAAGATCTCAGCCCAGGAGAGCCGTCGTAAGAAGAAGGAGTATGTGGAGTGTCTAGAAAAGAAGGTGGAGACATTTACATCTGAGAACAATGAACTGTGGAAGAAGGTGGAGACCCTGGAGAATGCCAACAGGACCCTGCTCCAGCAGCTGCAGAAA"</span>,
    CREB3L2 <span class="op">=</span> <span class="st">"GAGAAGGCCCTGAAGAAAATTCGGAGGAAGATCAAGAATAAGATTTCTGCTCAGGAAAGTAGGAGAAAGAAGAAAGAATACATGGACAGCCTGGAGAAAAAAGTGGAGTCTTGTTCAACTGAGAACTTGGAGCTTCGGAAGAAGGTAGAGGTTCTAGAGAACACTAATAGGACTCTCCTTCAGCAACTCCAGAAG"</span>,
    CREB3L3 <span class="op">=</span> <span class="st">"GAGCGAGTGCTGAAAAAAATCCGCCGGAAAATCCGGAACAAGCAGTCGGCGCAAGAAAGCAGGAAGAAGAAGAAGGAATATATCGATGGCCTGGAGACTCGGATGTCAGCTTGCACTGCTCAGAATCAGGAGTTACAGAGGAAAGTCTTGCATCTCGAGAAGCAAAACCTGTCCCTCTTGGAGCAACTGAAGAAA"</span>,
    CREB3L4 <span class="op">=</span> <span class="st">"GAGAGGGTCCTCAAGAAGGTCAGGAGGAAAATCCGTAACAAGCAGTCAGCTCAGGACAGTCGGCGGCGGAAGAAGGAGTACATTGATGGGCTGGAGAGCAGGGTGGCAGCCTGTTCTGCACAGAACCAAGAATTACAGAAAAAAGTCCAGGAGCTGGAGAGGCACAACATCTCCTTGGTAGCTCAGCTCCGCCAG"</span>,
    CREBL2 <span class="op">=</span> <span class="st">"CCAGCCAAAATTGACTTGAAAGCAAAACTTGAGAGGAGCCGGCAGAGTGCAAGAGAATGCCGAGCCCGAAAAAAGCTGAGATATCAGTATTTGGAAGAGTTGGTATCCAGTCGAGAAAGAGCTATATGTGCCCTCAGAGAGGAACTGGAAATGTACAAGCAGTGGTGCATGGCAATGGACCAAGGAAAAATCCCT"</span>,
    CREBRF <span class="op">=</span> <span class="st">"CCCTTAACAGCCCGACCAAGGTCAAGGAAGGAAAAAAATAAGCTGGCTTCCAGAGCTTGTCGGTTAAAGAAGAAAGCCCAGTATGAAGCTAATAAAGTGAAATTATGGGGCCTCAACACAGAATATGATAATTTATTGTTTGTAATCAACTCCATCAAGCAAGAGATTGTAAACCGGGTACAGAATCCAAGAGAT"</span>,
    DBP <span class="op">=</span> <span class="st">"CAGAAGGATGAGAAATACTGGAGCCGGCGGTACAAGAACAACGAGGCAGCCAAGCGGTCCCGTGACGCCCGGCGGCTCAAGGAGAACCAGATATCGGTGCGGGCGGCCTTCCTGGAGAAGGAGAACGCCCTGCTGCGGCAGGAAGTTGTGGCCGTGCGCCAGGAGCTGTCCCACTACCGCGCCGTGCTGTCCCGA"</span>,
    NFE2L2 <span class="op">=</span> <span class="st">"CTTGCATTAATTCGGGATATACGTAGGAGGGGTAAGAATAAAGTGGCTGCTCAGAATTGCAGAAAAAGAAAACTGGAAAATATAGTAGAACTAGAGCAAGATTTAGATCATTTGAAAGATGAAAAAGAAAAATTGCTCAAAGAAAAAGGAGAAAATGACAAAAGCCTTCACCTACTGAAAAAACAACTCAGCACC"</span>,
    NFE2L3 <span class="op">=</span> <span class="st">"GTCTCACTTATCCGTGACATCAGACGAAGAGGGAAAAATAAAGTTGCTGCGCAGAACTGTCGTAAACGCAAATTGGACATAATTTTGAATTTAGAAGATGATGTATGTAACTTGCAAGCAAAGAAGGAAACTCTTAAGAGAGAGCAAGCACAATGTAACAAAGCTATTAACATAATGAAACAGAAACTGCATGAC"</span>,
    TEF <span class="op">=</span> <span class="st">"CAGAAGGATGAAAAGTACTGGACAAGACGCAAGAAGAACAACGTGGCAGCTAAACGGTCACGGGATGCCCGGCGCCTGAAAGAGAATCAGATCACCATCCGGGCAGCCTTCCTGGAGAAGGAGAACACAGCCCTGCGGACGGAGGTGGCCGAGCTACGCAAGGAGGTGGGCAAGTGCAAGACCATCGTGTCCAAG"</span>
<span class="op">)</span></code></pre></div>
<p>Next we process the data, not allowing any mismatches in the forward read, but 1 mismatching codon in the reverse read. Now, we assume that the variable sequence starts immediately after the provided primers, and hence we don’t specify any UMI/constant sequence lengths. For the forward read, the variable region is taken to be the remainder of the read (after the primer), whereas for the reverse read, we specify the variable sequence length to 96.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">leujunt0</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/digestFastqs.html">digestFastqs</a></span><span class="op">(</span>
    fastqForward <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"leujunt0_1.fastq.gz"</span><span class="op">)</span>,
    fastqReverse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"leujunt0_2.fastq.gz"</span><span class="op">)</span>,
    mergeForwardReverse <span class="op">=</span> <span class="cn">FALSE</span>, 
    revComplForward <span class="op">=</span> <span class="cn">FALSE</span>, revComplReverse <span class="op">=</span> <span class="cn">FALSE</span>,
    elementsForward <span class="op">=</span> <span class="st">"SPV"</span>, elementsReverse <span class="op">=</span> <span class="st">"SPVS"</span>,
    elementLengthsForward <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">19</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span>,
    elementLengthsReverse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">20</span>, <span class="fl">96</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span>,
    constantForward <span class="op">=</span> <span class="st">""</span>, constantReverse <span class="op">=</span> <span class="st">""</span>, 
    adapterForward <span class="op">=</span> <span class="st">""</span>, adapterReverse <span class="op">=</span> <span class="st">""</span>, 
    primerForward <span class="op">=</span> <span class="st">"GTCAGGTGGAGGCGGATCC"</span>, 
    primerReverse <span class="op">=</span> <span class="st">"GAAAAAGGAAGCTGGAGAGA"</span>,
    avePhredMinForward <span class="op">=</span> <span class="fl">20</span>, avePhredMinReverse <span class="op">=</span> <span class="fl">20</span>, 
    variableNMaxForward <span class="op">=</span> <span class="fl">0</span>, variableNMaxReverse <span class="op">=</span> <span class="fl">0</span>, umiNMax <span class="op">=</span> <span class="fl">0</span>,
    wildTypeForward <span class="op">=</span> <span class="va">leu</span>, 
    wildTypeReverse <span class="op">=</span> <span class="st">"ATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTT"</span>, 
    nbrMutatedCodonsMaxForward <span class="op">=</span> <span class="fl">0</span>, 
    nbrMutatedCodonsMaxReverse <span class="op">=</span> <span class="fl">1</span>, 
    forbiddenMutatedCodonsForward <span class="op">=</span> <span class="st">""</span>,
    forbiddenMutatedCodonsReverse <span class="op">=</span> <span class="st">"NNW"</span>,
    mutatedPhredMinForward <span class="op">=</span> <span class="fl">0.0</span>, mutatedPhredMinReverse <span class="op">=</span> <span class="fl">0.0</span>,
    mutNameDelimiter <span class="op">=</span> <span class="st">"."</span>, 
    verbose <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>
<span class="co">#&gt; done enumerating forbidden codons (0)</span>
<span class="co">#&gt; done enumerating forbidden codons (32)</span>
<span class="co">#&gt; start reading sequences for file pair 1 of 1...</span>
<span class="co">#&gt;     1000 read pairs processed (60% retained)</span>
<span class="co">#&gt; done reading sequences</span>
<span class="co">#&gt; retained 587 unique features</span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">leujunt0</span><span class="op">$</span><span class="va">parameters</span>
<span class="co">#&gt; $fastqForward</span>
<span class="co">#&gt; [1] "/Users/runner/work/_temp/Library/mutscan/extdata/leujunt0_1.fastq.gz"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $fastqReverse</span>
<span class="co">#&gt; [1] "/Users/runner/work/_temp/Library/mutscan/extdata/leujunt0_2.fastq.gz"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $mergeForwardReverse</span>
<span class="co">#&gt; [1] FALSE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $minOverlap</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $maxOverlap</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $minMergedLength</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $maxMergedLength</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $maxFracMismatchOverlap</span>
<span class="co">#&gt; [1] 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $greedyOverlap</span>
<span class="co">#&gt; [1] TRUE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $revComplForward</span>
<span class="co">#&gt; [1] FALSE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $revComplReverse</span>
<span class="co">#&gt; [1] FALSE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $elementsForward</span>
<span class="co">#&gt; [1] "SPV"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $elementLengthsForward</span>
<span class="co">#&gt; [1] -1 19 -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $elementsReverse</span>
<span class="co">#&gt; [1] "SPVS"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $elementLengthsReverse</span>
<span class="co">#&gt; [1] -1 20 96 -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $adapterForward</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $adapterReverse</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $primerForward</span>
<span class="co">#&gt; [1] "GTCAGGTGGAGGCGGATCC"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $primerReverse</span>
<span class="co">#&gt; [1] "GAAAAAGGAAGCTGGAGAGA"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $wildTypeForward</span>
<span class="co">#&gt;                                                                                                                                                                                                  ATF2 </span>
<span class="co">#&gt; "GATCCTGATGAAAAAAGGAGAAAGTTTTTAGAGCGAAATAGAGCAGCAGCTTCAAGATGCCGACAAAAAAGGAAAGTCTGGGTTCAGTCTTTAGAGAAGAAAGCTGAAGACTTGAGTTCATTAAATGGTCAGCTGCAGAGTGAAGTCACCCTGCTGAGAAATGAAGTGGCACAGCTGAAACAGCTTCTTCTGGCT" </span>
<span class="co">#&gt;                                                                                                                                                                                                  ATF7 </span>
<span class="co">#&gt; "GATCCAGATGAGCGACGGCAGCGCTTTCTGGAGCGCAACCGGGCTGCAGCCTCCCGCTGCCGCCAAAAGCGAAAGCTGTGGGTGTCCTCCCTAGAGAAGAAGGCCGAAGAACTCACTTCTCAGAACATTCAGCTGAGTAATGAAGTCACATTACTACGCAATGAGGTGGCCCAGTTGAAACAGCTACTGTTAGCT" </span>
<span class="co">#&gt;                                                                                                                                                                                                 CREB5 </span>
<span class="co">#&gt; "GATCCGGACGAGAGGCGGCGGAAATTTCTGGAACGGAACCGGGCAGCTGCCACCCGCTGCAGACAGAAGAGGAAGGTCTGGGTGATGTCATTGGAAAAGAAAGCAGAAGAACTCACCCAGACAAACATGCAGCTTCAGAATGAAGTGTCTATGTTGAAAAATGAGGTGGCCCAGCTGAAACAGTTGTTGTTAACA" </span>
<span class="co">#&gt;                                                                                                                                                                                                  ATF3 </span>
<span class="co">#&gt; "GAAGAAGATGAAAGGAAAAAGAGGCGACGAGAAAGAAATAAGATTGCAGCTGCAAAGTGCCGAAACAAGAAGAAGGAGAAGACGGAGTGCCTGCAGAAAGAGTCGGAGAAGCTGGAAAGTGTGAATGCTGAACTGAAGGCTCAGATTGAGGAGCTCAAGAACGAGAAGCAGCATTTGATATACATGCTCAACCTT" </span>
<span class="co">#&gt;                                                                                                                                                                                                  JDP2 </span>
<span class="co">#&gt; "GAGGAAGAGGAGCGAAGGAAAAGGCGCCGGGAGAAGAACAAAGTCGCAGCAGCCCGATGCCGGAACAAGAAGAAGGAGCGCACGGAGTTTCTGCAGCGGGAATCCGAGCGGCTGGAACTCATGAACGCAGAGCTGAAGACCCAGATTGAGGAGCTGAAGCAGGAGCGGCAGCAGCTCATCCTGATGCTGAACCGA" </span>
<span class="co">#&gt;                                                                                                                                                                                                  ATF4 </span>
<span class="co">#&gt; "GAGAAACTGGATAAGAAGCTGAAAAAAATGGAGCAAAACAAGACAGCAGCCACTAGGTACCGCCAGAAGAAGAGGGCGGAGCAGGAGGCTCTTACTGGTGAGTGCAAAGAGCTGGAAAAGAAGAACGAGGCTCTAAAAGAGAGGGCGGATTCCCTGGCCAAGGAGATCCAGTACCTGAAAGATTTGATAGAAGAG" </span>
<span class="co">#&gt;                                                                                                                                                                                                  ATF5 </span>
<span class="co">#&gt; "ACCCGAGGGGACCGCAAGCAAAAGAAGAGAGACCAGAACAAGTCGGCGGCTCTGAGGTACCGCCAGCGGAAGCGGGCAGAGGGTGAGGCCCTGGAGGGCGAGTGCCAGGGGCTGGAGGCACGGAATCGCGAGCTGAAGGAACGGGCAGAGTCCGTGGAGCGCGAGATCCAGTACGTCAAGGACCTGCTCATCGAG" </span>
<span class="co">#&gt;                                                                                                                                                                                                CREBZF </span>
<span class="co">#&gt; "AGTCCCCGGAAGGCGGCGGCGGCCGCTGCCCGCCTTAATCGACTGAAGAAGAAGGAGTACGTGATGGGGCTGGAGAGTCGAGTCCGGGGTCTGGCAGCCGAGAACCAGGAGCTGCGGGCCGAGAATCGGGAGCTGGGCAAACGCGTACAGGCACTGCAGGAGGAGAGTCGCTACCTACGGGCAGTCTTAGCCAAC" </span>
<span class="co">#&gt;                                                                                                                                                                                                 BATF2 </span>
<span class="co">#&gt; "CCCAAGGAGCAACAAAGGCAGCTGAAGAAGCAGAAGAACCGGGCAGCCGCCCAGCGAAGCCGGCAGAAGCACACAGACAAGGCAGACGCCCTGCACCAGCAGCACGAGTCTCTGGAAAAAGACAACCTCGCCCTGCGGAAGGAGATCCAGTCCCTGCAGGCCGAGCTGGCGTGGTGGAGCCGGACCCTGCACGTG" </span>
<span class="co">#&gt;                                                                                                                                                                                                 BATF3 </span>
<span class="co">#&gt; "GAGGATGATGACAGGAAGGTCCGAAGGAGAGAAAAAAACCGAGTTGCTGCTCAGAGAAGTCGGAAGAAGCAGACCCAGAAGGCTGACAAGCTCCATGAGGAATATGAGAGCCTGGAGCAAGAAAACACCATGCTGCGGAGAGAGATCGGGAAGCTGACAGAGGAGCTGAAGCACCTGACAGAGGCACTGAAGGAG" </span>
<span class="co">#&gt;                                                                                                                                                                                                 CEBPE </span>
<span class="co">#&gt; "AAAGATAGCCTTGAGTACCGGCTGAGGCGGGAGCGCAACAACATCGCCGTGCGCAAGAGCCGAGACAAGGCCAAGAGGCGCATTCTGGAGACGCAGCAGAAGGTGCTGGAGTACATGGCAGAGAACGAGCGCCTCCGCAGCCGCGTGGAGCAGCTCACCCAGGAGCTAGACACCCTCCGCAACCTCTTCCGCCAG" </span>
<span class="co">#&gt;                                                                                                                                                                                                 BACH1 </span>
<span class="co">#&gt; "CTGGATTGTATCCATGATATTCGAAGAAGAAGTAAAAACAGAATTGCTGCACAGCGCTGTCGCAAGAGAAAACTTGACTGTATACAGAATCTTGAATCAGAAATTGAGAAGCTGCAAAGTGAAAAGGAGAGCTTGTTGAAGGAAAGAGATCACATTTTGTCAACTCTGGGTGAGACAAAGCAGAACCTAACTGGA" </span>
<span class="co">#&gt;                                                                                                                                                                                                 BACH2 </span>
<span class="co">#&gt; "TTAGAGTTTATTCATGATGTCCGACGGCGCAGCAAGAACCGCATCGCGGCCCAGCGCTGCCGCAAAAGGAAACTGGACTGTATTCAGAATTTAGAATGTGAAATCCGCAAATTGGTGTGTGAGAAAGAGAAACTGTTGTCAGAGAGGAATCAACTGAAAGCATGCATGGGGGAACTGTTGGACAACTTCTCCTGC" </span>
<span class="co">#&gt;                                                                                                                                                                                                NFE2L1 </span>
<span class="co">#&gt; "CTGAGCCTCATCCGAGACATCCGGCGCCGGGGCAAGAACAAGATGGCGGCGCAGAACTGCCGCAAGCGCAAGCTGGACACCATCCTGAATCTGGAGCGTGATGTGGAGGACCTGCAGCGTGACAAAGCCCGGCTGCTGCGGGAGAAAGTGGAGTTCCTGCGCTCCCTGCGACAGATGAAGCAGAAGGTCCAGAGC" </span>
<span class="co">#&gt;                                                                                                                                                                                                  NFE2 </span>
<span class="co">#&gt; "CTAGCGCTAGTCCGGGACATCCGACGACGGGGCAAAAACAAGGTGGCAGCCCAGAACTGCCGCAAGAGGAAGCTGGAAACCATTGTGCAGCTGGAGCGGGAGCTGGAGCGGCTGACCAATGAACGGGAGCGGCTTCTCAGGGCCCGCGGGGAGGCAGACCGGACCCTGGAGGTCATGCGCCAACAGCTGACAGAG" </span>
<span class="co">#&gt;                                                                                                                                                                                                 NFIL3 </span>
<span class="co">#&gt; "AAGAAAGATGCTATGTATTGGGAAAAAAGGCGGAAAAATAATGAAGCTGCCAAAAGATCTCGTGAGAAGCGTCGACTGAATGACCTGGTTTTAGAGAACAAACTAATTGCACTGGGAGAAGAAAACGCCACTTTAAAAGCTGAGCTGCTTTCACTAAAATTAAAGTTTGGTTTAATTAGCTCCACAGCATATGCT" </span>
<span class="co">#&gt;                                                                                                                                                                                                   FOS </span>
<span class="co">#&gt; "GAAGAAGAAGAGAAAAGGAGAATCCGAAGGGAAAGGAATAAGATGGCTGCAGCCAAATGCCGCAACCGGAGGAGGGAGCTGACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTAGAGTTCATCCTGGCAGCT" </span>
<span class="co">#&gt;                                                                                                                                                                                                  FOSB </span>
<span class="co">#&gt; "GAGGAAGAGGAGAAGCGAAGGGTGCGCCGGGAACGAAATAAACTAGCAGCAGCTAAATGCAGGAACCGGCGGAGGGAGCTGACCGACCGACTCCAGGCGGAGACAGATCAGTTGGAGGAAGAAAAAGCAGAGCTGGAGTCGGAGATCGCCGAGCTCCAAAAGGAGAAGGAACGTCTGGAGTTTGTGCTGGTGGCC" </span>
<span class="co">#&gt;                                                                                                                                                                                                 FOSL1 </span>
<span class="co">#&gt; "GAGGAAGAGGAGCGCCGCCGAGTAAGGCGCGAGCGGAACAAGCTGGCTGCGGCCAAGTGCAGGAACCGGAGGAAGGAACTGACCGACTTCCTGCAGGCGGAGACTGACAAACTGGAAGATGAGAAATCTGGGCTGCAGCGAGAGATTGAGGAGCTGCAGAAGCAGAAGGAGCGCCTAGAGCTGGTGCTGGAAGCC" </span>
<span class="co">#&gt;                                                                                                                                                                                                 FOSL2 </span>
<span class="co">#&gt; "GAAGAGGAGGAGAAGCGTCGCATCCGGCGGGAGAGGAACAAGCTGGCTGCAGCCAAGTGCCGGAACCGACGCCGGGAGCTGACAGAGAAGCTGCAGGCGGAGACAGAGGAGCTGGAGGAGGAGAAGTCAGGCCTGCAGAAGGAGATTGCTGAGCTGCAGAAGGAGAAGGAGAAGCTGGAGTTCATGTTGGTGGCT" </span>
<span class="co">#&gt;                                                                                                                                                                                                  MAFB </span>
<span class="co">#&gt; "GTGATCCGCCTGAAGCAGAAGCGGCGGACCCTGAAGAACCGGGGCTACGCCCAGTCTTGCAGGTATAAACGCGTCCAGCAGAAGCACCACCTGGAGAATGAGAAGACGCAGCTCATTCAGCAGGTGGAGCAGCTTAAGCAGGAGGTGTCCCGGCTGGCCCGCGAGAGAGACGCCTACAAGGTCAAGTGCGAGAAA" </span>
<span class="co">#&gt;                                                                                                                                                                                                   JUN </span>
<span class="co">#&gt; "CAGGAGCGGATCAAGGCGGAGAGGAAGCGCATGAGGAACCGCATCGCTGCCTCCAAGTGCCGAAAAAGGAAGCTGGAGAGAATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTTAAACAGAAAGTCATGAAC" </span>
<span class="co">#&gt;                                                                                                                                                                                                  JUNB </span>
<span class="co">#&gt; "CAAGAGCGCATCAAAGTGGAGCGCAAGCGGCTGCGGAACCGGCTGGCGGCCACCAAGTGCCGGAAGCGGAAGCTGGAGCGCATCGCGCGCCTGGAGGACAAGGTGAAGACGCTCAAGGCCGAGAACGCGGGGCTGTCGAGTACCGCCGGCCTCCTCCGGGAGCAGGTGGCCCAGCTCAAACAGAAGGTCATGACC" </span>
<span class="co">#&gt;                                                                                                                                                                                                  JUND </span>
<span class="co">#&gt; "CAGGAGCGCATCAAGGCGGAGCGCAAGCGGCTGCGCAACCGCATCGCCGCCTCCAAGTGCCGCAAGCGCAAGCTGGAGCGCATCTCGCGCCTGGAAGAGAAAGTGAAGACCCTCAAGAGTCAGAACACGGAGCTGGCGTCCACGGCGAGCCTGCTGCGCGAGCAGGTGGCGCAGCTCAAGCAGAAAGTCCTCAGC" </span>
<span class="co">#&gt;                                                                                                                                                                                                 CREB3 </span>
<span class="co">#&gt; "GAACAAATTCTGAAACGTGTGCGGAGGAAGATTCGAAATAAAAGATCTGCTCAAGAGAGCCGCAGGAAAAAGAAGGTGTATGTTGGGGGTTTAGAGAGCAGGGTCTTGAAATACACAGCCCAGAATATGGAGCTTCAGAACAAAGTACAGCTTCTGGAGGAACAGAATTTGTCCCTTCTAGATCAACTGAGGAAA" </span>
<span class="co">#&gt;                                                                                                                                                                                                   HLF </span>
<span class="co">#&gt; "CTGAAGGATGACAAGTACTGGGCAAGGCGCAGAAAGAACAACATGGCAGCCAAGCGCTCCCGCGACGCCCGGAGGCTGAAAGAGAACCAGATCGCCATCCGGGCCTCGTTCCTGGAGAAGGAGAACTCGGCCCTCCGCCAGGAGGTGGCTGACTTGAGGAAGGAGCTGGGCAAATGCAAGAACATACTTGCCAAG" </span>
<span class="co">#&gt;                                                                                                                                                                                                  MAFG </span>
<span class="co">#&gt; "ATCGTCCAGCTGAAGCAGCGCCGGCGCACGCTCAAGAACCGCGGCTACGCTGCCAGCTGCCGCGTGAAGCGGGTGACGCAGAAGGAGGAGCTGGAGAAGCAGAAGGCGGAGCTGCAGCAGGAGGTGGAGAAGCTGGCCTCAGAGAACGCCAGCATGAAGCTGGAGCTCGACGCGCTGCGCTCCAAGTACGAGGCG" </span>
<span class="co">#&gt;                                                                                                                                                                                                  MAFK </span>
<span class="co">#&gt; "GTGACCCGCCTGAAGCAGCGTCGGCGCACACTCAAGAACCGCGGCTACGCGGCCAGCTGCCGCATCAAGCGGGTGACGCAGAAGGAGGAGCTGGAGCGGCAGCGCGTGGAGCTGCAGCAGGAGGTGGAGAAGCTGGCGCGTGAGAACAGCAGCATGCGGCTGGAGCTGGACGCCCTGCGCTCCAAGTACGAGGCG" </span>
<span class="co">#&gt;                                                                                                                                                                                                  XBP1 </span>
<span class="co">#&gt; "AGCCCCGAGGAGAAGGCGCTGAGGAGGAAACTGAAAAACAGAGTAGCAGCTCAGACTGCCAGAGATCGAAAGAAGGCTCGAATGAGTGAGCTGGAACAGCAAGTGGTAGATTTAGAAGAAGAGAACCAAAAACTTTTGCTAGAAAATCAGCTTTTACGAGAGAAAACTCATGGCCTTGTAGTTGAGAACCAGGAG" </span>
<span class="co">#&gt;                                                                                                                                                                                                  ATF6 </span>
<span class="co">#&gt; "ATTGCTGTGCTAAGGAGACAGCAACGTATGATAAAAAATCGAGAATCCGCTTGTCAGTCTCGCAAGAAGAAGAAAGAATATATGCTAGGGTTAGAGGCGAGATTAAAGGCTGCCCTCTCAGAAAACGAGCAACTGAAGAAAGAAAATGGAACACTGAAGCGGCAGCTGGATGAAGTTGTGTCAGAGAACCAGAGG" </span>
<span class="co">#&gt;                                                                                                                                                                                                 ATF6B </span>
<span class="co">#&gt; "GCAAAGCTGCTGAAGCGGCAGCAGCGAATGATCAAGAACCGGGAGTCAGCCTGCCAGTCCCGGAGAAAGAAGAAAGAGTATCTGCAGGGACTGGAGGCTCGGCTGCAAGCAGTACTGGCTGACAACCAGCAGCTCCGCCGAGAGAATGCTGCCCTCCGGCGGCGGCTGGAGGCCCTGCTGGCTGAAAACAGCGAG" </span>
<span class="co">#&gt;                                                                                                                                                                                                 CEBPA </span>
<span class="co">#&gt; "AAGAACAGCAACGAGTACCGGGTGCGGCGCGAGCGCAACAACATCGCGGTGCGCAAGAGCCGCGACAAGGCCAAGCAGCGCAACGTGGAGACGCAGCAGAAGGTGCTGGAGCTGACCAGTGACAATGACCGCCTGCGCAAGCGGGTGGAACAGCTGAGCCGCGAACTGGACACGCTGCGGGGCATCTTCCGCCAG" </span>
<span class="co">#&gt;                                                                                                                                                                                                 CEBPB </span>
<span class="co">#&gt; "AAGCACAGCGACGAGTACAAGATCCGGCGCGAGCGCAACAACATCGCCGTGCGCAAGAGCCGCGACAAGGCCAAGATGCGCAACCTGGAGACGCAGCACAAGGTCCTGGAGCTCACGGCCGAGAACGAGCGGCTGCAGAAGAAGGTGGAGCAGCTGTCGCGCGAGCTCAGCACCCTGCGGAACTTGTTCAAGCAG" </span>
<span class="co">#&gt;                                                                                                                                                                                                 CEBPD </span>
<span class="co">#&gt; "CGCGGCAGCCCCGAGTACCGGCAGCGGCGCGAGCGCAACAACATCGCCGTGCGCAAGAGCCGCGACAAGGCCAAGCGGCGCAACCAGGAGATGCAGCAGAAGTTGGTGGAGCTGTCGGCTGAGAACGAGAAGCTGCACCAGCGCGTGGAGCAGCTCACGCGGGACCTGGCCGGCCTCCGGCAGTTCTTCAAGCAG" </span>
<span class="co">#&gt;                                                                                                                                                                                                 CEBPG </span>
<span class="co">#&gt; "CGAAACAGTGACGAGTATCGGCAACGCCGAGAGAGGAACAACATGGCTGTGAAAAAGAGCCGGTTGAAAAGCAAGCAGAAAGCACAAGACACACTGCAGAGAGTCAATCAGCTCAAAGAAGAGAATGAACGGTTGGAAGCAAAAATCAAATTGCTGACCAAGGAATTAAGTGTACTCAAAGATTTGTTTCTTGAG" </span>
<span class="co">#&gt;                                                                                                                                                                                                 CREB1 </span>
<span class="co">#&gt;             "GAAGCAGCACGAAAGAGAGAGGTCCGTCTAATGAAGAACAGGGAAGCAGCTCGAGAGTGTCGTAGAAAGAAGAAAGAATATGTGAAATGTTTAGAAAACAGAGTGGCAGTGCTTGAAAATCAAAACAAGACATTGATTGAGGAGCTAAAAGCACTTAAGGACCTTTACTGCCACAAATCAGAT" </span>
<span class="co">#&gt;                                                                                                                                                                                               CREB3L1 </span>
<span class="co">#&gt; "GAGAAGGCCTTGAAGAGAGTCCGGAGGAAAATCAAGAACAAGATCTCAGCCCAGGAGAGCCGTCGTAAGAAGAAGGAGTATGTGGAGTGTCTAGAAAAGAAGGTGGAGACATTTACATCTGAGAACAATGAACTGTGGAAGAAGGTGGAGACCCTGGAGAATGCCAACAGGACCCTGCTCCAGCAGCTGCAGAAA" </span>
<span class="co">#&gt;                                                                                                                                                                                               CREB3L2 </span>
<span class="co">#&gt; "GAGAAGGCCCTGAAGAAAATTCGGAGGAAGATCAAGAATAAGATTTCTGCTCAGGAAAGTAGGAGAAAGAAGAAAGAATACATGGACAGCCTGGAGAAAAAAGTGGAGTCTTGTTCAACTGAGAACTTGGAGCTTCGGAAGAAGGTAGAGGTTCTAGAGAACACTAATAGGACTCTCCTTCAGCAACTCCAGAAG" </span>
<span class="co">#&gt;                                                                                                                                                                                               CREB3L3 </span>
<span class="co">#&gt; "GAGCGAGTGCTGAAAAAAATCCGCCGGAAAATCCGGAACAAGCAGTCGGCGCAAGAAAGCAGGAAGAAGAAGAAGGAATATATCGATGGCCTGGAGACTCGGATGTCAGCTTGCACTGCTCAGAATCAGGAGTTACAGAGGAAAGTCTTGCATCTCGAGAAGCAAAACCTGTCCCTCTTGGAGCAACTGAAGAAA" </span>
<span class="co">#&gt;                                                                                                                                                                                               CREB3L4 </span>
<span class="co">#&gt; "GAGAGGGTCCTCAAGAAGGTCAGGAGGAAAATCCGTAACAAGCAGTCAGCTCAGGACAGTCGGCGGCGGAAGAAGGAGTACATTGATGGGCTGGAGAGCAGGGTGGCAGCCTGTTCTGCACAGAACCAAGAATTACAGAAAAAAGTCCAGGAGCTGGAGAGGCACAACATCTCCTTGGTAGCTCAGCTCCGCCAG" </span>
<span class="co">#&gt;                                                                                                                                                                                                CREBL2 </span>
<span class="co">#&gt; "CCAGCCAAAATTGACTTGAAAGCAAAACTTGAGAGGAGCCGGCAGAGTGCAAGAGAATGCCGAGCCCGAAAAAAGCTGAGATATCAGTATTTGGAAGAGTTGGTATCCAGTCGAGAAAGAGCTATATGTGCCCTCAGAGAGGAACTGGAAATGTACAAGCAGTGGTGCATGGCAATGGACCAAGGAAAAATCCCT" </span>
<span class="co">#&gt;                                                                                                                                                                                                CREBRF </span>
<span class="co">#&gt; "CCCTTAACAGCCCGACCAAGGTCAAGGAAGGAAAAAAATAAGCTGGCTTCCAGAGCTTGTCGGTTAAAGAAGAAAGCCCAGTATGAAGCTAATAAAGTGAAATTATGGGGCCTCAACACAGAATATGATAATTTATTGTTTGTAATCAACTCCATCAAGCAAGAGATTGTAAACCGGGTACAGAATCCAAGAGAT" </span>
<span class="co">#&gt;                                                                                                                                                                                                   DBP </span>
<span class="co">#&gt; "CAGAAGGATGAGAAATACTGGAGCCGGCGGTACAAGAACAACGAGGCAGCCAAGCGGTCCCGTGACGCCCGGCGGCTCAAGGAGAACCAGATATCGGTGCGGGCGGCCTTCCTGGAGAAGGAGAACGCCCTGCTGCGGCAGGAAGTTGTGGCCGTGCGCCAGGAGCTGTCCCACTACCGCGCCGTGCTGTCCCGA" </span>
<span class="co">#&gt;                                                                                                                                                                                                NFE2L2 </span>
<span class="co">#&gt; "CTTGCATTAATTCGGGATATACGTAGGAGGGGTAAGAATAAAGTGGCTGCTCAGAATTGCAGAAAAAGAAAACTGGAAAATATAGTAGAACTAGAGCAAGATTTAGATCATTTGAAAGATGAAAAAGAAAAATTGCTCAAAGAAAAAGGAGAAAATGACAAAAGCCTTCACCTACTGAAAAAACAACTCAGCACC" </span>
<span class="co">#&gt;                                                                                                                                                                                                NFE2L3 </span>
<span class="co">#&gt; "GTCTCACTTATCCGTGACATCAGACGAAGAGGGAAAAATAAAGTTGCTGCGCAGAACTGTCGTAAACGCAAATTGGACATAATTTTGAATTTAGAAGATGATGTATGTAACTTGCAAGCAAAGAAGGAAACTCTTAAGAGAGAGCAAGCACAATGTAACAAAGCTATTAACATAATGAAACAGAAACTGCATGAC" </span>
<span class="co">#&gt;                                                                                                                                                                                                   TEF </span>
<span class="co">#&gt; "CAGAAGGATGAAAAGTACTGGACAAGACGCAAGAAGAACAACGTGGCAGCTAAACGGTCACGGGATGCCCGGCGCCTGAAAGAGAATCAGATCACCATCCGGGCAGCCTTCCTGGAGAAGGAGAACACAGCCCTGCGGACGGAGGTGGCCGAGCTACGCAAGGAGGTGGGCAAGTGCAAGACCATCGTGTCCAAG" </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $wildTypeReverse</span>
<span class="co">#&gt;                                                                                                  r </span>
<span class="co">#&gt; "ATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTT" </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $constantForward</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $constantReverse</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $avePhredMinForward</span>
<span class="co">#&gt; [1] 20</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $avePhredMinReverse</span>
<span class="co">#&gt; [1] 20</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $variableNMaxForward</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $variableNMaxReverse</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $umiNMax</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $nbrMutatedCodonsMaxForward</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $nbrMutatedCodonsMaxReverse</span>
<span class="co">#&gt; [1] 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $nbrMutatedBasesMaxForward</span>
<span class="co">#&gt; [1] -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $nbrMutatedBasesMaxReverse</span>
<span class="co">#&gt; [1] -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $forbiddenMutatedCodonsForward</span>
<span class="co">#&gt; character(0)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $forbiddenMutatedCodonsReverse</span>
<span class="co">#&gt;  [1] "AAA" "AAT" "ACA" "ACT" "AGA" "AGT" "ATA" "ATT" "CAA" "CAT" "CCA" "CCT"</span>
<span class="co">#&gt; [13] "CGA" "CGT" "CTA" "CTT" "GAA" "GAT" "GCA" "GCT" "GGA" "GGT" "GTA" "GTT"</span>
<span class="co">#&gt; [25] "TAA" "TAT" "TCA" "TCT" "TGA" "TGT" "TTA" "TTT"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $useTreeWTmatch</span>
<span class="co">#&gt; [1] FALSE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $mutatedPhredMinForward</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $mutatedPhredMinReverse</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $mutNameDelimiter</span>
<span class="co">#&gt; [1] "."</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $constantMaxDistForward</span>
<span class="co">#&gt; [1] -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $constantMaxDistReverse</span>
<span class="co">#&gt; [1] -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $variableCollapseMaxDist</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $variableCollapseMinReads</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $variableCollapseMinRatio</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $umiCollapseMaxDist</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $filteredReadsFastqForward</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $filteredReadsFastqReverse</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $maxNReads</span>
<span class="co">#&gt; [1] -1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $nThreads</span>
<span class="co">#&gt; [1] 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $chunkSize</span>
<span class="co">#&gt; [1] 100000</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $processingInfo</span>
<span class="co">#&gt; [1] "Processed by mutscan v0.2.26 on 2022-01-21 14:31:54"</span>
<span class="va">leujunt0</span><span class="op">$</span><span class="va">filterSummary</span>
<span class="co">#&gt;   nbrTotal f1_nbrAdapter f2_nbrNoPrimer f3_nbrReadWrongLength</span>
<span class="co">#&gt; 1     1000             0            126                     0</span>
<span class="co">#&gt;   f4_nbrNoValidOverlap f5_nbrAvgVarQualTooLow f6_nbrTooManyNinVar</span>
<span class="co">#&gt; 1                    0                      0                  76</span>
<span class="co">#&gt;   f7_nbrTooManyNinUMI f8_nbrTooManyBestWTHits f9_nbrMutQualTooLow</span>
<span class="co">#&gt; 1                   0                       0                   0</span>
<span class="co">#&gt;   f10a_nbrTooManyMutCodons f10b_nbrTooManyMutBases f11_nbrForbiddenCodons</span>
<span class="co">#&gt; 1                      195                       0                      3</span>
<span class="co">#&gt;   f12_nbrTooManyMutConstant f13_nbrTooManyBestConstantHits nbrRetained</span>
<span class="co">#&gt; 1                         0                              0         600</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">leujunt0</span><span class="op">$</span><span class="va">summaryTable</span><span class="op">)</span>
<span class="co">#&gt;           mutantName</span>
<span class="co">#&gt; 1 ATF2.0.WT_r.13.CCC</span>
<span class="co">#&gt; 2 ATF2.0.WT_r.20.TGG</span>
<span class="co">#&gt; 3 ATF2.0.WT_r.32.CGG</span>
<span class="co">#&gt; 4  ATF2.0.WT_r.4.ATG</span>
<span class="co">#&gt; 5  ATF2.0.WT_r.5.ATG</span>
<span class="co">#&gt; 6  ATF2.0.WT_r.6.GAG</span>
<span class="co">#&gt;                                                                                                                                                                                                     sequence</span>
<span class="co">#&gt; 1 GATCCTGATGAAAAAAGGAGAAAGTTTTTAGAGCGAAATAGAGCAGCAGCTTCAAGATGCCGACAAAAAAGGAAAGTCTGGGTTCAGTCTTTAGAGAAGAAAGCT_ATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAACCCCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTT</span>
<span class="co">#&gt; 2   GATCCTGATGAAAAAAGGAGAAAGTTTTTAGAGCGAAATAGAGCAGCAGCTTCAAGATGCCGACAAAAAAGGAAAGTCTGGGTTCAGTCTTTAGAGAAGAAAG_ATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTGGACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTT</span>
<span class="co">#&gt; 3  GATCCTGATGAAAAAAGGAGAAAGTTTTTAGAGCGAAATAGAGCAGCAGCTTCAAGATGCCGACAAAAAAGGAAAGTCTGGGTTCAGTCTTTAGAGAAGAAAGC_ATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCGG</span>
<span class="co">#&gt; 4  GATCCTGATGAAAAAAGGAGAAAGTTTTTAGAGCGAAATAGAGCAGCAGCTTCAAGATGCCGACAAAAAAGGAAAGTCTGGGTTCAGTCTTTAGAGAAGAAAGC_ATCGCCCGGATGGAGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTT</span>
<span class="co">#&gt; 5 GATCCTGATGAAAAAAGGAGAAAGTTTTTAGAGCGAAATAGAGCAGCAGCTTCAAGATGCCGACAAAAAAGGAAAGTCTGGGTTCAGTCTTTAGAGAAGAAAGCT_ATCGCCCGGCTGATGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTT</span>
<span class="co">#&gt; 6 GATCCTGATGAAAAAAGGAGAAAGTTTTTAGAGCGAAATAGAGCAGCAGCTTCAAGATGCCGACAAAAAAGGAAAGTCTGGGTTCAGTCTTTAGAGAAGAAAGCT_ATCGCCCGGCTGGAGGAGAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTT</span>
<span class="co">#&gt;   nbrMutBases nbrMutCodons nbrReads maxNbrReads nbrUmis</span>
<span class="co">#&gt; 1           2            1        1           1       0</span>
<span class="co">#&gt; 2           2            1        1           1       0</span>
<span class="co">#&gt; 3           2            1        1           1       0</span>
<span class="co">#&gt; 4           1            1        1           1       0</span>
<span class="co">#&gt; 5           2            1        1           1       0</span>
<span class="co">#&gt; 6           1            1        1           1       0</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="combining-samples">Combining samples<a class="anchor" aria-label="anchor" href="#combining-samples"></a>
</h2>
<p>The <code><a href="../reference/digestFastqs.html">digestFastqs()</a></code> function processes each sample (pair of FASTQ files) separately. In order to prepare the data for downstream statistical analysis and plotting, the <code><a href="../reference/summarizeExperiment.html">summarizeExperiment()</a></code> function takes a named list of outputs from <code><a href="../reference/digestFastqs.html">digestFastqs()</a></code>, and a <code>data.frame</code> with sample annotations, and generates a , with an <code>assay</code> containing either UMI or read counts. To illustrate this, we process also the output sample for the TRANS experiment for which we processed the input sample above, and feed both samples to <code><a href="../reference/summarizeExperiment.html">summarizeExperiment()</a></code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">transOutput</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/digestFastqs.html">digestFastqs</a></span><span class="op">(</span>
    fastqForward <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"transOutput_1.fastq.gz"</span><span class="op">)</span>,
    fastqReverse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"transOutput_2.fastq.gz"</span><span class="op">)</span>,
    mergeForwardReverse <span class="op">=</span> <span class="cn">FALSE</span>, 
    revComplForward <span class="op">=</span> <span class="cn">FALSE</span>, revComplReverse <span class="op">=</span> <span class="cn">FALSE</span>,
    adapterForward <span class="op">=</span> <span class="st">"GGAAGAGCACACGTC"</span>, 
    adapterReverse <span class="op">=</span> <span class="st">"GGAAGAGCGTCGTGT"</span>, 
    elementsForward <span class="op">=</span> <span class="st">"SUCV"</span>, elementsReverse <span class="op">=</span> <span class="st">"SUCV"</span>,
    elementLengthsForward <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10</span>, <span class="fl">18</span>, <span class="fl">96</span><span class="op">)</span>,
    elementLengthsReverse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">8</span>, <span class="fl">20</span>, <span class="fl">96</span><span class="op">)</span>,
    constantForward <span class="op">=</span> <span class="st">"AACCGGAGGAGGGAGCTG"</span>,
    constantReverse <span class="op">=</span> <span class="st">"GAAAAAGGAAGCTGGAGAGA"</span>, 
    primerForward <span class="op">=</span> <span class="st">""</span>, primerReverse <span class="op">=</span> <span class="st">""</span>,
    avePhredMinForward <span class="op">=</span> <span class="fl">20</span>, avePhredMinReverse <span class="op">=</span> <span class="fl">20</span>,
    variableNMaxForward <span class="op">=</span> <span class="fl">0</span>, variableNMaxReverse <span class="op">=</span> <span class="fl">0</span>, umiNMax <span class="op">=</span> <span class="fl">0</span>,
    wildTypeForward <span class="op">=</span> <span class="st">"ACTGATACACTCCAAGCGGAGACAGACCAACTAGAAGATGAGAAGTCTGCTTTGCAGACCGAGATTGCCAACCTGCTGAAGGAGAAGGAAAAACTA"</span>, 
    wildTypeReverse <span class="op">=</span> <span class="st">"ATCGCCCGGCTGGAGGAAAAAGTGAAAACCTTGAAAGCTCAGAACTCGGAGCTGGCGTCCACGGCCAACATGCTCAGGGAACAGGTGGCACAGCTT"</span>, 
    nbrMutatedCodonsMaxForward <span class="op">=</span> <span class="fl">1</span>, nbrMutatedCodonsMaxReverse <span class="op">=</span> <span class="fl">1</span>, 
    forbiddenMutatedCodonsForward <span class="op">=</span> <span class="st">"NNW"</span>, forbiddenMutatedCodonsReverse <span class="op">=</span> <span class="st">"NNW"</span>,
    mutNameDelimiter <span class="op">=</span> <span class="st">"."</span>, 
    constantMaxDistForward <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,
    constantMaxDistReverse <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,
    verbose <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>

<span class="co">## Generate SummarizedExperiment object</span>
<span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarizeExperiment.html">summarizeExperiment</a></span><span class="op">(</span>
    x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>sample1 <span class="op">=</span> <span class="va">transInput</span>,
             sample2 <span class="op">=</span> <span class="va">transOutput</span><span class="op">)</span>,
    coldata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span>,
                         Condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"input"</span>, <span class="st">"output"</span><span class="op">)</span>,
                         Replicate <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"R1"</span>, <span class="st">"R1"</span><span class="op">)</span>,
                         OD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="fl">1.5</span><span class="op">)</span><span class="op">)</span>
<span class="op">)</span>

<span class="co">## The SummarizedExperiment contains a count matrix, and annotations </span>
<span class="co">## for samples and variants in the `colData` and `rowData`, </span>
<span class="co">## respectively.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; 6 x 2 sparse Matrix of class "dgTMatrix"</span>
<span class="co">#&gt;                 sample1 sample2</span>
<span class="co">#&gt; f.0.WT_r.13.CCC       1       .</span>
<span class="co">#&gt; f.0.WT_r.13.CTC       1       .</span>
<span class="co">#&gt; f.0.WT_r.2.TCC        1       .</span>
<span class="co">#&gt; f.0.WT_r.20.ACC       1       .</span>
<span class="co">#&gt; f.0.WT_r.30.AGG       1       .</span>
<span class="co">#&gt; f.0.WT_r.30.GGG       1       .</span>
<span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; sample1 sample2 </span>
<span class="co">#&gt;     279     285</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; DataFrame with 6 rows and 8 columns</span>
<span class="co">#&gt;                      mutantName               sequence   nbrMutBases</span>
<span class="co">#&gt;                     &lt;character&gt;            &lt;character&gt; &lt;IntegerList&gt;</span>
<span class="co">#&gt; f.0.WT_r.13.CCC f.0.WT_r.13.CCC ACTGATACACTCCAAGCGGA..             2</span>
<span class="co">#&gt; f.0.WT_r.13.CTC f.0.WT_r.13.CTC ACTGATACACTCCAAGCGGA..             3</span>
<span class="co">#&gt; f.0.WT_r.2.TCC   f.0.WT_r.2.TCC ACTGATACACTCCAAGCGGA..             1</span>
<span class="co">#&gt; f.0.WT_r.20.ACC f.0.WT_r.20.ACC ACTGATACACTCCAAGCGGA..             1</span>
<span class="co">#&gt; f.0.WT_r.30.AGG f.0.WT_r.30.AGG ACTGATACACTCCAAGCGGA..             3</span>
<span class="co">#&gt; f.0.WT_r.30.GGG f.0.WT_r.30.GGG ACTGATACACTCCAAGCGGA..             2</span>
<span class="co">#&gt;                  nbrMutCodons minNbrMutBases maxNbrMutBases minNbrMutCodons</span>
<span class="co">#&gt;                 &lt;IntegerList&gt;      &lt;integer&gt;      &lt;integer&gt;       &lt;integer&gt;</span>
<span class="co">#&gt; f.0.WT_r.13.CCC             1              2              2               1</span>
<span class="co">#&gt; f.0.WT_r.13.CTC             1              3              3               1</span>
<span class="co">#&gt; f.0.WT_r.2.TCC              1              1              1               1</span>
<span class="co">#&gt; f.0.WT_r.20.ACC             1              1              1               1</span>
<span class="co">#&gt; f.0.WT_r.30.AGG             1              3              3               1</span>
<span class="co">#&gt; f.0.WT_r.30.GGG             1              2              2               1</span>
<span class="co">#&gt;                 maxNbrMutCodons</span>
<span class="co">#&gt;                       &lt;integer&gt;</span>
<span class="co">#&gt; f.0.WT_r.13.CCC               1</span>
<span class="co">#&gt; f.0.WT_r.13.CTC               1</span>
<span class="co">#&gt; f.0.WT_r.2.TCC                1</span>
<span class="co">#&gt; f.0.WT_r.20.ACC               1</span>
<span class="co">#&gt; f.0.WT_r.30.AGG               1</span>
<span class="co">#&gt; f.0.WT_r.30.GGG               1</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>
<span class="co">#&gt; DataFrame with 2 rows and 20 columns</span>
<span class="co">#&gt;                Name   Condition   Replicate        OD  nbrTotal f1_nbrAdapter</span>
<span class="co">#&gt;         &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;numeric&gt; &lt;integer&gt;     &lt;integer&gt;</span>
<span class="co">#&gt; sample1     sample1       input          R1      0.05      1000           314</span>
<span class="co">#&gt; sample2     sample2      output          R1      1.50      1000           366</span>
<span class="co">#&gt;         f2_nbrNoPrimer f3_nbrReadWrongLength f4_nbrNoValidOverlap</span>
<span class="co">#&gt;              &lt;integer&gt;             &lt;integer&gt;            &lt;integer&gt;</span>
<span class="co">#&gt; sample1              0                     0                    0</span>
<span class="co">#&gt; sample2              0                     0                    0</span>
<span class="co">#&gt;         f5_nbrAvgVarQualTooLow f6_nbrTooManyNinVar f7_nbrTooManyNinUMI</span>
<span class="co">#&gt;                      &lt;integer&gt;           &lt;integer&gt;           &lt;integer&gt;</span>
<span class="co">#&gt; sample1                      7                   0                   0</span>
<span class="co">#&gt; sample2                      6                   0                   0</span>
<span class="co">#&gt;         f8_nbrTooManyBestWTHits f9_nbrMutQualTooLow f10a_nbrTooManyMutCodons</span>
<span class="co">#&gt;                       &lt;integer&gt;           &lt;integer&gt;                &lt;integer&gt;</span>
<span class="co">#&gt; sample1                       0                   0                      392</span>
<span class="co">#&gt; sample2                       0                   0                      338</span>
<span class="co">#&gt;         f10b_nbrTooManyMutBases f11_nbrForbiddenCodons</span>
<span class="co">#&gt;                       &lt;integer&gt;              &lt;integer&gt;</span>
<span class="co">#&gt; sample1                       0                      8</span>
<span class="co">#&gt; sample2                       0                      5</span>
<span class="co">#&gt;         f12_nbrTooManyMutConstant f13_nbrTooManyBestConstantHits nbrRetained</span>
<span class="co">#&gt;                         &lt;integer&gt;                      &lt;integer&gt;   &lt;integer&gt;</span>
<span class="co">#&gt; sample1                         0                              0         279</span>
<span class="co">#&gt; sample2                         0                              0         285</span>

<span class="co">## Count type (reads or UMIs)</span>
<span class="fu">metadata</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">countType</span>
<span class="co">#&gt; [1] "umis"</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="collapsing-count-matrix-to-amino-acids">Collapsing count matrix to amino acids<a class="anchor" aria-label="anchor" href="#collapsing-count-matrix-to-amino-acids"></a>
</h2>
<p>The  object generated by <code><a href="../reference/summarizeExperiment.html">summarizeExperiment()</a></code> contains one row for each observed variant (combination). This can be further collapsed by replacing the mutated codon by the corresponding amino acid, and aggregating the counts corresponding to the same mutated amino acid (combination).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se_collapsed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/collapseMutantsByAA.html">collapseMutantsByAA</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_collapsed</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; 6 x 2 sparse Matrix of class "dgCMatrix"</span>
<span class="co">#&gt;               sample1 sample2</span>
<span class="co">#&gt; f.0.NA_r.0.NA       .       1</span>
<span class="co">#&gt; f.0.NA_r.13.A       .       1</span>
<span class="co">#&gt; f.0.NA_r.13.L       1       .</span>
<span class="co">#&gt; f.0.NA_r.13.P       1       .</span>
<span class="co">#&gt; f.0.NA_r.2.S        1       .</span>
<span class="co">#&gt; f.0.NA_r.20.T       1       .</span>
<span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_collapsed</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; sample1 sample2 </span>
<span class="co">#&gt;     279     285</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se_collapsed</span><span class="op">)</span>
<span class="co">#&gt; DataFrame with 2 rows and 20 columns</span>
<span class="co">#&gt;                Name   Condition   Replicate        OD  nbrTotal f1_nbrAdapter</span>
<span class="co">#&gt;         &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;numeric&gt; &lt;integer&gt;     &lt;integer&gt;</span>
<span class="co">#&gt; sample1     sample1       input          R1      0.05      1000           314</span>
<span class="co">#&gt; sample2     sample2      output          R1      1.50      1000           366</span>
<span class="co">#&gt;         f2_nbrNoPrimer f3_nbrReadWrongLength f4_nbrNoValidOverlap</span>
<span class="co">#&gt;              &lt;integer&gt;             &lt;integer&gt;            &lt;integer&gt;</span>
<span class="co">#&gt; sample1              0                     0                    0</span>
<span class="co">#&gt; sample2              0                     0                    0</span>
<span class="co">#&gt;         f5_nbrAvgVarQualTooLow f6_nbrTooManyNinVar f7_nbrTooManyNinUMI</span>
<span class="co">#&gt;                      &lt;integer&gt;           &lt;integer&gt;           &lt;integer&gt;</span>
<span class="co">#&gt; sample1                      7                   0                   0</span>
<span class="co">#&gt; sample2                      6                   0                   0</span>
<span class="co">#&gt;         f8_nbrTooManyBestWTHits f9_nbrMutQualTooLow f10a_nbrTooManyMutCodons</span>
<span class="co">#&gt;                       &lt;integer&gt;           &lt;integer&gt;                &lt;integer&gt;</span>
<span class="co">#&gt; sample1                       0                   0                      392</span>
<span class="co">#&gt; sample2                       0                   0                      338</span>
<span class="co">#&gt;         f10b_nbrTooManyMutBases f11_nbrForbiddenCodons</span>
<span class="co">#&gt;                       &lt;integer&gt;              &lt;integer&gt;</span>
<span class="co">#&gt; sample1                       0                      8</span>
<span class="co">#&gt; sample2                       0                      5</span>
<span class="co">#&gt;         f12_nbrTooManyMutConstant f13_nbrTooManyBestConstantHits nbrRetained</span>
<span class="co">#&gt;                         &lt;integer&gt;                      &lt;integer&gt;   &lt;integer&gt;</span>
<span class="co">#&gt; sample1                         0                              0         279</span>
<span class="co">#&gt; sample2                         0                              0         285</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="diagnostic-plots">Diagnostic plots<a class="anchor" aria-label="anchor" href="#diagnostic-plots"></a>
</h2>
<p><code>mutscan</code> contains functionality for generating a variety of diagnostic plots. Here we illustrate these using the full CIS data set from <span class="citation">[@Diss2018]</span>, which has been processed using <code><a href="../reference/digestFastqs.html">digestFastqs()</a></code> as illustrated above, and summarized in a <code>SummarizedExperiment</code> object provided with the package.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"GSE102901_cis_se.rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>First, we can plot a summary of the filtering process, indicating the number of reads that were filtered out by (or retained after) each step of the <code>mutscan</code> filtering.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotFiltering.html">plotFiltering</a></span><span class="op">(</span><span class="va">se</span>, valueType <span class="op">=</span> <span class="st">"reads"</span>, onlyActiveFilters <span class="op">=</span> <span class="cn">TRUE</span>, 
              plotType <span class="op">=</span> <span class="st">"remaining"</span>, facetBy <span class="op">=</span> <span class="st">"sample"</span>, numberSize <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="mutscan_files/figure-html/unnamed-chunk-16-1.png" width="768"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotFiltering.html">plotFiltering</a></span><span class="op">(</span><span class="va">se</span>, valueType <span class="op">=</span> <span class="st">"fractions"</span>, onlyActiveFilters <span class="op">=</span> <span class="cn">TRUE</span>,
              plotType <span class="op">=</span> <span class="st">"filtered"</span>, facetBy <span class="op">=</span> <span class="st">"step"</span>, numberSize <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="mutscan_files/figure-html/unnamed-chunk-17-1.png" width="768"></p>
<p>We can also generate a pairs plot displaying the correlation among the samples in the data set.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotPairs.html">plotPairs</a></span><span class="op">(</span><span class="va">se</span>, selAssay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></code></pre></div>
<p><img src="mutscan_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>Additional plot functions can be used to visualize the total count per sample, across all variants, or the distribution of variant counts per sample.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotTotals.html">plotTotals</a></span><span class="op">(</span><span class="va">se</span>, selAssay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></code></pre></div>
<p><img src="mutscan_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotDistributions.html">plotDistributions</a></span><span class="op">(</span><span class="va">se</span>, selAssay <span class="op">=</span> <span class="st">"counts"</span>, plotType <span class="op">=</span> <span class="st">"density"</span>, 
                  pseudocount <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="mutscan_files/figure-html/unnamed-chunk-19-2.png" width="700"></p>
<p>Finally, we can create a full QC report as follows:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/generateQCReport.html">generateQCReport</a></span><span class="op">(</span><span class="va">se</span>, outFile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">"html"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="calculating-ppi-scores">Calculating PPI scores<a class="anchor" aria-label="anchor" href="#calculating-ppi-scores"></a>
</h2>
<p>The function <code><a href="../reference/calculatePPIScore.html">calculatePPIScore()</a></code> can be used to calculate PPI scores as described in <span class="citation">[@Diss2018]</span>. The function requires the user to specify a <code>pairingCol</code>, containing the replicate ID for each sample; one or more <code>ODCols</code>, containing the optical density for each sample, and a <code>comparison</code>, which is a character vector of length 3 specifying the comparison to perform, of the form (<code>groupColumn</code>, <code>numerator</code>, <code>denominator</code>). Here, <code>groupColumn</code> is the name of the column in <code>colData(se)</code> that contains the grouping information, and <code>numerator</code> and <code>denominator</code> specify the values of this column representing the two groups to be compared.</p>
<p>Here, we illustrate the application of <code><a href="../reference/calculatePPIScore.html">calculatePPIScore()</a></code> on the SummarizedExperiment object containing all the three CIS replicates from <span class="citation">[@Diss2018]</span>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se_collapsed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/collapseMutantsByAA.html">collapseMutantsByAA</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>
<span class="va">ppis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculatePPIScore.html">calculatePPIScore</a></span><span class="op">(</span>se <span class="op">=</span> <span class="va">se_collapsed</span>, pairingCol <span class="op">=</span> <span class="st">"Replicate"</span>, 
                          ODCols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"OD1"</span>, <span class="st">"OD2"</span><span class="op">)</span>,
                          comparison <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Condition"</span>, <span class="st">"cis_output"</span>, <span class="st">"cis_input"</span><span class="op">)</span>,
                          WTrows <span class="op">=</span> <span class="st">"f.0.NA"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ppis</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">ppis</span><span class="op">)</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;        cis_output_vs_cis_input_repl1 cis_output_vs_cis_input_repl2</span>
<span class="co">#&gt; f.8.M                       1.078839                      1.077029</span>
<span class="co">#&gt; f.8.A                       1.077068                      1.087273</span>
<span class="co">#&gt; f.2.Q                       1.093031                      1.065204</span>
<span class="co">#&gt; f.3.M                       1.083608                      1.077612</span>
<span class="co">#&gt; f.9.K                       1.085180                      1.037331</span>
<span class="co">#&gt; f.13.K                      1.042529                      1.045749</span>
<span class="co">#&gt;        cis_output_vs_cis_input_repl3</span>
<span class="co">#&gt; f.8.M                       1.116606</span>
<span class="co">#&gt; f.8.A                       1.102605</span>
<span class="co">#&gt; f.2.Q                       1.064832</span>
<span class="co">#&gt; f.3.M                       1.050167</span>
<span class="co">#&gt; f.9.K                       1.061854</span>
<span class="co">#&gt; f.13.K                      1.094965</span>

<span class="co">## The PPI score for the WT sequence is 1, by construction</span>
<span class="va">ppis</span><span class="op">[</span><span class="st">"f.0.NA"</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>
<span class="co">#&gt;        cis_output_vs_cis_input_repl1 cis_output_vs_cis_input_repl2</span>
<span class="co">#&gt; f.0.NA                             1                             1</span>
<span class="co">#&gt;        cis_output_vs_cis_input_repl3</span>
<span class="co">#&gt; f.0.NA                             1</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="scoring-mutations-with-edger-or-limma">Scoring mutations with <code>edgeR</code> or <code>limma</code><a class="anchor" aria-label="anchor" href="#scoring-mutations-with-edger-or-limma"></a>
</h2>
<p>As an alternative to the PPI scoring, <code>mutscan</code> can be used to model the observed counts using a generalized linear model (with <code>edgeR</code>) or a general linear model (with <code>limma</code>) and estimate a logFC and a p-value for the enrichment of each variant betwen two conditions (or more generally, in association with any predictor), compared to one or more WT sequences. Note that for this, at least two replicates are required.</p>
<p>We start by looking at the design matrix, in order to determine which of the coefficients to specify for the testing in <code><a href="../reference/calculateRelativeFC.html">calculateRelativeFC()</a></code>. For more information about how to set up and interpret design matrices in <code>edgeR</code> or <code>limma</code>, see e.g. <a href="https://f1000research.com/articles/9-1444" class="external-link">Law et al (2020)</a>, or <a href="https://f1000research.com/articles/9-512" class="external-link">Soneson et al (2020)</a>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">Replicate</span> <span class="op">+</span> <span class="va">Condition</span>,
             data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se_collapsed</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;            (Intercept) Replicate2 Replicate3 Conditioncis_output</span>
<span class="co">#&gt; SRR5952435           1          0          0                   0</span>
<span class="co">#&gt; SRR5952436           1          1          0                   0</span>
<span class="co">#&gt; SRR5952437           1          0          1                   0</span>
<span class="co">#&gt; SRR5952438           1          0          0                   1</span>
<span class="co">#&gt; SRR5952439           1          1          0                   1</span>
<span class="co">#&gt; SRR5952440           1          0          1                   1</span>
<span class="co">#&gt; attr(,"assign")</span>
<span class="co">#&gt; [1] 0 1 1 2</span>
<span class="co">#&gt; attr(,"contrasts")</span>
<span class="co">#&gt; attr(,"contrasts")$Replicate</span>
<span class="co">#&gt; [1] "contr.treatment"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attr(,"contrasts")$Condition</span>
<span class="co">#&gt; [1] "contr.treatment"</span></code></pre></div>
<p>Next, we apply either <code>edgeR</code> or <code>limma</code> to extract the logFCs of the mutants, compared to the wildtype sequence.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## edgeR</span>
<span class="va">edger_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateRelativeFC.html">calculateRelativeFC</a></span><span class="op">(</span>
  se <span class="op">=</span> <span class="va">se_collapsed</span>,
  design <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">Replicate</span> <span class="op">+</span> <span class="va">Condition</span>,
                        data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se_collapsed</span><span class="op">)</span><span class="op">)</span>,
  coef <span class="op">=</span> <span class="st">"Conditioncis_output"</span>, pseudocount <span class="op">=</span> <span class="fl">1</span>, WTrows <span class="op">=</span> <span class="st">"f.0.NA"</span>,
  method <span class="op">=</span> <span class="st">"edgeR"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">edger_scores</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">edger_scores</span><span class="op">$</span><span class="va">PValue</span><span class="op">)</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;            logFC   logCPM        F       PValue          FDR logFC_shrunk</span>
<span class="co">#&gt; f.27.* -4.508548 11.12488 21848.85 8.040788e-23 2.396903e-20    -4.507885</span>
<span class="co">#&gt; f.7.G  -4.189336 11.14507 21833.58 8.078333e-23 2.396903e-20    -4.188817</span>
<span class="co">#&gt; f.28.G -3.950089 11.21692 20884.26 1.086210e-22 2.396903e-20    -3.949669</span>
<span class="co">#&gt; f.28.* -4.392247 11.14109 19422.47 1.761245e-22 2.914860e-20    -4.391639</span>
<span class="co">#&gt; f.7.*  -4.465149 10.87741 18497.72 2.437401e-22 3.227119e-20    -4.464383</span>
<span class="co">#&gt; f.14.* -4.378119 10.90439 17510.04 3.512726e-22 3.742515e-20    -4.377415</span>
<span class="co">## As before, the WT sequence has a logFC close to 0, by construction</span>
<span class="va">edger_scores</span><span class="op">[</span><span class="st">"f.0.NA"</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>
<span class="co">#&gt;                logFC   logCPM             F PValue FDR logFC_shrunk</span>
<span class="co">#&gt; f.0.NA -4.230076e-15 18.99263 -1.549382e-08      1   1 1.061987e-15</span>

<span class="co">## limma</span>
<span class="va">limma_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateRelativeFC.html">calculateRelativeFC</a></span><span class="op">(</span>
  se <span class="op">=</span> <span class="va">se_collapsed</span>,
  design <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">Replicate</span> <span class="op">+</span> <span class="va">Condition</span>,
                        data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se_collapsed</span><span class="op">)</span><span class="op">)</span>,
  coef <span class="op">=</span> <span class="st">"Conditioncis_output"</span>, pseudocount <span class="op">=</span> <span class="fl">1</span>, WTrows <span class="op">=</span> <span class="st">"f.0.NA"</span>,
  method <span class="op">=</span> <span class="st">"limma"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">limma_scores</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">limma_scores</span><span class="op">$</span><span class="va">P.Value</span><span class="op">)</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;            logFC      CI.L      CI.R   AveExpr         t      P.Value</span>
<span class="co">#&gt; f.4.F  -2.901012 -2.972503 -2.829522 11.116487 -83.86175 1.550530e-30</span>
<span class="co">#&gt; f.28.G -3.949985 -4.048503 -3.851466 10.298829 -82.85952 2.052780e-30</span>
<span class="co">#&gt; f.7.G  -4.189243 -4.295813 -4.082673 10.124214 -81.23885 3.254876e-30</span>
<span class="co">#&gt; f.27.* -4.508302 -4.624500 -4.392103  9.963169 -80.18203 4.417988e-30</span>
<span class="co">#&gt; f.28.* -4.392002 -4.505390 -4.278614 10.031157 -80.04944 4.591925e-30</span>
<span class="co">#&gt; f.4.V  -3.391374 -3.480293 -3.302456 10.866682 -78.82166 6.585796e-30</span>
<span class="co">#&gt;           adj.P.Val        B   se.logFC</span>
<span class="co">#&gt; f.4.F  6.079709e-28 60.12115 0.03459279</span>
<span class="co">#&gt; f.28.G 6.079709e-28 59.82096 0.04767086</span>
<span class="co">#&gt; f.7.G  6.079709e-28 59.32763 0.05156699</span>
<span class="co">#&gt; f.27.* 6.079709e-28 58.97200 0.05622583</span>
<span class="co">#&gt; f.28.* 6.079709e-28 58.96010 0.05486612</span>
<span class="co">#&gt; f.4.V  6.200353e-28 58.69311 0.04302592</span>
<span class="co">## As before, the WT sequence has a logFC close to 0, by construction</span>
<span class="va">limma_scores</span><span class="op">[</span><span class="st">"f.0.NA"</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>
<span class="co">#&gt;               logFC         CI.L        CI.R AveExpr            t   P.Value</span>
<span class="co">#&gt; f.0.NA 4.426619e-09 -0.008386242 0.008386251 18.9826 1.090858e-06 0.9999991</span>
<span class="co">#&gt;        adj.P.Val         B    se.logFC</span>
<span class="co">#&gt; f.0.NA 0.9999991 -11.27198 0.004057922</span></code></pre></div>
<p><code>mutscan</code> also contains functions for plotting the results of the statistical testing - in particular, MA (mean-difference) plots and volcano plots can be easily generated.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotMeanDiff.html">plotMeanDiff</a></span><span class="op">(</span><span class="va">edger_scores</span>, pointSize <span class="op">=</span> <span class="st">"large"</span><span class="op">)</span></code></pre></div>
<p><img src="mutscan_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotVolcano.html">plotVolcano</a></span><span class="op">(</span><span class="va">edger_scores</span>, pointSize <span class="op">=</span> <span class="st">"large"</span><span class="op">)</span></code></pre></div>
<p><img src="mutscan_files/figure-html/unnamed-chunk-24-2.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>This vignette was compiled on the following system:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Big Sur 10.16</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">#&gt; [8] base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] mutscan_0.2.26              SummarizedExperiment_1.24.0</span>
<span class="co">#&gt;  [3] Biobase_2.54.0              GenomicRanges_1.46.1       </span>
<span class="co">#&gt;  [5] GenomeInfoDb_1.30.0         IRanges_2.28.0             </span>
<span class="co">#&gt;  [7] S4Vectors_0.32.3            BiocGenerics_0.40.0        </span>
<span class="co">#&gt;  [9] MatrixGenerics_1.6.0        matrixStats_0.61.0         </span>
<span class="co">#&gt; [11] BiocStyle_2.22.0           </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] tidyr_1.1.4            edgeR_3.36.0           sass_0.4.0            </span>
<span class="co">#&gt;  [4] splines_4.1.2          jsonlite_1.7.3         bslib_0.3.1           </span>
<span class="co">#&gt;  [7] highr_0.9              BiocManager_1.30.16    GenomeInfoDbData_1.2.7</span>
<span class="co">#&gt; [10] Rsamtools_2.10.0       ggrepel_0.9.1          yaml_2.2.1            </span>
<span class="co">#&gt; [13] pillar_1.6.4           lattice_0.20-45        limma_3.50.0          </span>
<span class="co">#&gt; [16] glue_1.6.0             digest_0.6.29          RColorBrewer_1.1-2    </span>
<span class="co">#&gt; [19] XVector_0.34.0         colorspace_2.0-2       htmltools_0.5.2       </span>
<span class="co">#&gt; [22] Matrix_1.4-0           plyr_1.8.6             pkgconfig_2.0.3       </span>
<span class="co">#&gt; [25] csaw_1.28.0            bookdown_0.24          zlibbioc_1.40.0       </span>
<span class="co">#&gt; [28] purrr_0.3.4            scales_1.1.1           BiocParallel_1.28.3   </span>
<span class="co">#&gt; [31] tibble_3.1.6           farver_2.1.0           generics_0.1.1        </span>
<span class="co">#&gt; [34] grr_0.9.5              ggplot2_3.3.5          ellipsis_0.3.2        </span>
<span class="co">#&gt; [37] DT_0.20                cachem_1.0.6           magrittr_2.0.1        </span>
<span class="co">#&gt; [40] crayon_1.4.2           memoise_2.0.1          evaluate_0.14         </span>
<span class="co">#&gt; [43] GGally_2.1.2           fs_1.5.2               fansi_1.0.2           </span>
<span class="co">#&gt; [46] textshaping_0.3.6      tools_4.1.2            lifecycle_1.0.1       </span>
<span class="co">#&gt; [49] stringr_1.4.0          locfit_1.5-9.4         munsell_0.5.0         </span>
<span class="co">#&gt; [52] DelayedArray_0.20.0    Biostrings_2.62.0      compiler_4.1.2        </span>
<span class="co">#&gt; [55] pkgdown_2.0.2.9000     jquerylib_0.1.4        systemfonts_1.0.3     </span>
<span class="co">#&gt; [58] rlang_0.4.12           grid_4.1.2             Matrix.utils_0.9.8    </span>
<span class="co">#&gt; [61] RCurl_1.98-1.5         htmlwidgets_1.5.4      labeling_0.4.2        </span>
<span class="co">#&gt; [64] bitops_1.0-7           rmarkdown_2.11         gtable_0.3.0          </span>
<span class="co">#&gt; [67] reshape_0.8.8          R6_2.5.1               knitr_1.37            </span>
<span class="co">#&gt; [70] dplyr_1.0.7            fastmap_1.1.0          utf8_1.2.2            </span>
<span class="co">#&gt; [73] rprojroot_2.0.2        ragg_1.2.1             metapod_1.2.0         </span>
<span class="co">#&gt; [76] desc_1.4.0             stringi_1.7.6          parallel_4.1.2        </span>
<span class="co">#&gt; [79] Rcpp_1.0.8             vctrs_0.3.8            tidyselect_1.1.1      </span>
<span class="co">#&gt; [82] xfun_0.29</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Charlotte Soneson, Michael Stadler.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
