#' Collapse counts by mutated amino acid
#'
#' @param se A \code{\link[SummarizedExperiment]{SummarizedExperiment}}
#'     generated by \code{\link{summarizeExperiment}}
#' @param nameCol A character scalar providing the column of 
#'     \code{rowData(se)} that contains the amino acid mutant names (that will 
#'     be the new row names).
#'
#' @return A \code{\link[SummarizedExperiment]{SummarizedExperiment}} where
#'     counts have been aggregated by the mutated amino acid(s).
#'
#' @author Charlotte Soneson, Michael Stadler
#'
#' @export
#'
#' @importFrom DelayedArray rowsum
#' @importFrom S4Vectors metadata
#' @importFrom SummarizedExperiment assays rowData SummarizedExperiment colData
#' @importFrom dplyr across group_by summarize
#' @importFrom stats setNames
#'
#' @examples 
#' se <- readRDS(system.file("extdata", "GSE102901_cis_se.rds",
#'                           package = "mutscan"))[1:200, ]
#' ## The rows of this object correspond to individual codon variants
#' dim(se)
#' head(rownames(se))
#' ## Collapse by amino acid
#' sec <- collapseMutantsByAA(se)
#' ## The rows of the collapsed object correspond to amino acid variants
#' dim(sec)
#' head(rownames(sec))
#' ## The mutantName column contains the individual codon variants that were 
#' ## collapsed
#' head(SummarizedExperiment::rowData(sec))
#' 
collapseMutantsByAA <- function(se, nameCol = "mutantNameAA") {
    .assertVector(x = se, type = "SummarizedExperiment")
    .assertScalar(x = nameCol, type = "character",
                  validValues = colnames(SummarizedExperiment::rowData(se)))
    
    ## Collapse assays
    ## Matrix.utils has been removed from CRAN
    # aList <- lapply(SummarizedExperiment::assays(se), function(a) {
    #     Matrix.utils::aggregate.Matrix(
    #         x = a,
    #         groupings = factor(SummarizedExperiment::rowData(se)[[nameCol]]),
    #         fun = "colSums")
    # })
    aList <- lapply(SummarizedExperiment::assays(se), function(a) {
        DelayedArray::rowsum(
            x = a,
            group = factor(SummarizedExperiment::rowData(se)[[nameCol]])
        )
    })
    
    ## Collapse rowData - simple columns
    rd <- mergeValues(SummarizedExperiment::rowData(se)[[nameCol]],
                      SummarizedExperiment::rowData(se)$sequence) %>%
        stats::setNames(c(nameCol, "sequence"))
    for (v in setdiff(
        intersect(c("mutantName", "mutantNameBase", "mutantNameBaseHGVS",
                    "mutantNameCodon", "mutantNameAA", "mutantNameAAHGVS",
                    "sequenceAA", "mutationTypes",
                    "nbrMutBases", "nbrMutCodons", "nbrMutAAs"),
                  colnames(SummarizedExperiment::rowData(se))),
        nameCol)) {
        tmp <- mergeValues(SummarizedExperiment::rowData(se)[[nameCol]],
                           SummarizedExperiment::rowData(se)[[v]])
        rd[[v]] <- tmp$valueColl[match(rd[[nameCol]], tmp$mutantNameColl)]
    }
    rd0 <- as.data.frame(SummarizedExperiment::rowData(se)) %>%
        dplyr::group_by(.data[[nameCol]]) %>%
        dplyr::summarize(
            dplyr::across(c("minNbrMutBases", "minNbrMutCodons",
                            "minNbrMutAAs"),
                          function(x) min(x)),
            dplyr::across(c("maxNbrMutBases", "maxNbrMutCodons",
                            "maxNbrMutAAs"),
                          function(x) max(x)))
    rd <- rd %>% 
        dplyr::full_join(rd0, by = nameCol)
    
    rd <- S4Vectors::DataFrame(rd)
    rownames(rd) <- rd[[nameCol]]
    
    ## Create SummarizedExperiment
    rn1 <- rownames(aList[[1]])
    cn1 <- colnames(aList[[1]])
    for (i in seq_along(aList)) {
        stopifnot(all(rownames(aList[[i]]) == rn1))
        stopifnot(all(colnames(aList[[i]]) == cn1))
    }
    stopifnot(cn1 == rownames(SummarizedExperiment::colData(se)))
    SummarizedExperiment::SummarizedExperiment(
        assays = aList,
        colData = SummarizedExperiment::colData(se),
        metadata = S4Vectors::metadata(se),
        rowData = rd[rn1, ]
    )
}
