# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @title Read, filter and digest sequences from two fastq files.
#'
#' @description
#' \code{readAndDigestFastqs} reads sequences for a pair of fastq files
#' and digests them (extracts umis, constant and variable parts, filters,
#' extracts mismatch information from constant and counts the observed
#' unique variable parts).
#'
#' @details
#' more details.
#'
#' @param experimentType character(1), either "cis" or "trans". If this is set
#'   to "cis", the variable sequences from the forward and reverse reads will be
#'   consolidated into one single sequence.
#' @param fastqForward,fastqReverse character(1), paths to FASTQ files
#'   corresponding to forward and reverse reads, respectively.
#' @param skipForward,skipReverse numeric(1), the number of bases to skip in the
#'   start of each forward and reverse read, respectively.
#' @param umiLengthForward,umiLengthReverse numeric(1), the length of the
#'   barcode (UMI) sequence in the forward/reverse reads, respectively, not
#'   including the skipped bases (defined by
#'   \code{skipForward}/\code{skipReverse}).
#' @param constantLengthForward,constantLengthReverse numeric(1), the length of
#'   the constant sequence in the forward/reverse reads, respectively.
#' @param variableLengthForward,variableLengthReverse numeric(1), the length of
#'   the variable sequence in the forward/reverse reads, respectively.
#' @param adapterForward,adapterReverse character(1), the adapter sequence for
#'   forward/reverse reads, respectively. If a forward/reverse read contains the
#'   corresponding adapter sequence, the sequence pair will be filtered out.
#'   If set to \code{NULL}, no adapter filtering is performed. The number of
#'   filtered read pairs are reported in the return value.
#' @param wildTypeForward,wildTypeReverse character(1), the wild type sequence
#'   for the forward and reverse variable region.
#' @param constantForward,constantReverse character(1), the expected constant
#'   forward and reverse sequences.
#' @param avePhredMin numeric(1) Minimum average Phred score in the variable
#'   region for a read to be retained. If L contains both forward and reverse
#'   variable regions, the minimum average Phred score has to be achieved in
#'   both for a read pair to be retained.
#' @param variableNMax numeric(1) Maximum number of Ns allowed in the variable
#'   region for a read to be retained.
#' @param umiNMax numeric(1) Maximum number of Ns allowed in the UMI for a read
#'   to be retained.
#' @param nbrMutatedCodonsMax numeric(1) Maximum number of mutated codons that
#'   are allowed.
#' @param forbiddenMutatedCodons character vector. Codons (can contain ambiguous
#'   IUPAC characters, see \code{\link[Biostrings]{IUPAC_CODE_MAP}}). If a read
#'   pair contains a mutated codon matching this pattern, it will be filtered
#'   out.
#' @param mutatedPhredMin numeric(1) Minimum Phred score of a mutated base for the
#'   read to be retained. If any mutated base has a Phred score lower than
#'   \code{mutatedPhredMin}, the read will be discarded.
#' @param verbose logical(1), whether to print out progress messages.
#'
#' @return A list with ---more details---.
#'
#' @export
digestFastqs <- function(experimentType, fastqForward, fastqReverse, skipForward = 1L, skipReverse = 1L, umiLengthForward = 10L, umiLengthReverse = 8L, constantLengthForward = 18L, constantLengthReverse = 20L, variableLengthForward = 96L, variableLengthReverse = 96L, adapterForward = "", adapterReverse = "", wildTypeForward = "", wildTypeReverse = "", constantForward = "", constantReverse = "", avePhredMin = 20.0, variableNMax = 0L, umiNMax = 0L, nbrMutatedCodonsMax = 1L, forbiddenMutatedCodons = "NNW", mutatedPhredMin = 0.0, verbose = FALSE) {
    .Call(`_mutscan_digestFastqs`, experimentType, fastqForward, fastqReverse, skipForward, skipReverse, umiLengthForward, umiLengthReverse, constantLengthForward, constantLengthReverse, variableLengthForward, variableLengthReverse, adapterForward, adapterReverse, wildTypeForward, wildTypeReverse, constantForward, constantReverse, avePhredMin, variableNMax, umiNMax, nbrMutatedCodonsMax, forbiddenMutatedCodons, mutatedPhredMin, verbose)
}

qual2int <- function(x) {
    .Call(`_mutscan_qual2int`, x)
}

